---
title: "Deprecated"
author: "Dan"
date: "9/17/2017"
output: html_document
---


#### Un-indexed

```{r}
# vs titration
std_kd %>% 
  mutate(protein = ifelse(protein == "pho4", "Pho4", "Cbf1")) %>% 
  select(X01:X10, protein, stdE, stdE.err) %>% 
  left_join(.,model_outputs %>% 
  tbl_df() %>% 
  select(X01:X10, .fitted, protein, model)) %>%
  mutate(model = factor(model, labels = c(expression(bold(paste("Mononucleotide"))),
                                          expression(bold(paste("Nearest neighbor"))),
                                          expression(bold(paste("Dinucleotide"))))),
         protein = factor(protein, 
                          levels = c("Pho4","Cbf1"),
                          labels = c(expression(bold(paste("Pho4"))),
                                     expression(bold(paste("Cbf1")))))) %>% 
  group_by(model,protein) %>% 
  summarise(r2 = cor(stdE, .fitted)^2) %>% 
  xtable() %>% 
  print()
```



## Figure S6: Raw vs. titration correlation
```{r}
crossallDat = 
  counts.df %>% 
  select(protein, flank, ddG) %>%
  left_join(std_kd %>% select(flank, stdE, stdE.err),.) %>% 
  mutate(protein = ifelse(protein == "pho4","Pho4","Cbf1"),
         protein = factor(protein, 
                          levels = c("Pho4","Cbf1")))

# plot
S6 = ggplot(crossallDat, aes(stdE,ddG)) +
  presentation +
  geom_errorbarh(aes(x = stdE, xmax = stdE + stdE.err, xmin = stdE - stdE.err), size = 0.25) +
  geom_point(size = 0.01) +
  theme(aspect.ratio = 1) +
  ylab(expression(bold(paste("Unprocessed ", Delta, Delta, "G, kcal/mol")))) +
  xlab(expression(bold(paste("Titration ", Delta, "G, kcal/mol")))) +
  facet_grid(protein ~ .)

ggsave("~/Flanks/Images/S6.png", plot = S6, height = 7, width = 7, units = "cm")
```

## Figure S8: Dinucleotides vs titration correlation
```{r,fig.height=10, fig.width=10}
predCor = 
  model_outputs %>% 
  tbl_df() %>% 
  select(X01:X10, .fitted, protein, model)

plotOut = 
  std_kd %>% 
  mutate(protein = ifelse(protein == "pho4", "Pho4", "Cbf1")) %>% 
  select(X01:X10, protein, stdE, stdE.err) %>% 
  left_join(.,predCor)

scaleDat = 
  plotOut %>%
  mutate(model = factor(model, labels = c(expression(bold(paste("Mononucleotide"))),
                                          expression(bold(paste("Nearest neighbor"))),
                                          expression(bold(paste("Dinucleotide"))))),
         protein = factor(protein, 
                          levels = c("Pho4","Cbf1"),
                          labels = c(expression(bold(paste("Pho4"))),
                                     expression(bold(paste("Cbf1"))))))

# plot
S8 = ggplot(scaleDat, aes(stdE,.fitted)) +
  presentation +
  geom_errorbarh(aes(x = stdE, xmin = stdE + stdE.err, xmax = stdE - stdE.err), size = 0.25) +
  geom_point(size = 0.01) +
  theme(aspect.ratio = 1) +
  facet_grid(protein ~ model, scales = "free", labeller = "label_parsed") +
  ylab(expression(bold(paste("Linear model ", Delta,Delta, "G, kcal/mol")))) +
  xlab(expression(bold(paste("Titration ", Delta, "G, kcal/mol"))))

ggsave("~/Flanks/Images/S8.png", plot = S8, height = 8, width = 8, units = "cm")


# correlation of linear model predictions vs titration
plotOut %>% 
  group_by(model,protein) %>% 
  summarise(r2 = cor(stdE, .fitted)^2) %>% 
  xtable() %>% 
  print()

# correlation of linear model preditions vs NN-derived ddG
corOut = 
  left_join(predCor %>% unite("flank",X01:X10,sep="") %>% select(flank,.fitted,protein,model), countsNN_scaled %>% select(flank,scaled_ddG,protein) %>% mutate(protein = ifelse(protein == "pho4", "Pho4", "Cbf1")))

corOut %>% 
  group_by(protein, model) %>% 
  summarise(r2 = cor(.fitted,scaled_ddG)^2) %>%
  xtable() %>% 
  print()
  
```





## Figure S12: Epistasis Pho4 titration binding curves
```{r}
colPal = colorRampPalette(c("#0057E7", "#D62D20"))(32)

pho4EpiCurves = 
  fread("~/Flank_seq/data/pho4EpistasisCurves.csv", sep = ",", header = T) %>%
  tbl_df() %>%
  rename(ID = Onum, fBound = R, Conc = Ch) %>%
  mutate(index = paste("ID",ID,sep=""), 
         index = factor(index, paste0("ID",1:32))) %>%
  createCurves(.) %>%
  mutate(index = factor(index, levels = paste0("ID",1:32))) %>%
  left_join(.,fread("~/Flank_seq/data/epi2ID_pho4.txt", header = F) %>%
    unite("flank",V2,V4,sep="_") %>%
    mutate(index = paste0("ID",1:32)) %>%
    select(index, flank)) %>%
  mutate(flank = ifelse(index == "ID32","negative control",flank)) %>%
  mutate(flank = factor(flank, levels = arrange(.,estimate) %>% .$flank %>% unique())) %>%
  filter(index != "ID10") # remove replicate

S12 = ggplot(pho4EpiCurves) +
  geom_point(aes(Conc,fBound,color=flank), size = 0.01) +
  geom_line(aes(Conc,.fitted,color=flank), size = 0.2) +
  presentation +
  theme(aspect.ratio = 1, 
        legend.position = "none",
        axis.text.x = element_text(angle = 90, vjust = 0.5),
        panel.spacing = unit(0.2, "lines")) +
  ylab("Intensity ratio (DNA/protein)") +
  xlab("DNA concentration, nM") + 
  facet_wrap( ~ flank, nrow=6) +
  scale_color_manual(values = colPal)

ggsave("~/Flanks/Images/S12.png", plot = S12, height = 17, width = 13, units = "cm")
```

## Figure S13: Pho4 dinucleotide LASSO benchmark
```{r}
dfVar = fread("~/Flank_seq/data/dfVar_pho4.txt", header =T)

# sum of additive features
stdEpi =
  fread("~/Flank_seq/data/stdEpistasis1.csv", header = T) %>%
  tbl_df() %>%
  mutate(Sequence = gsub(" ","",Sequence)) %>% # remove white space
  mutate(Sequence = substring(Sequence,16,31), Sequence = gsub("CACGTG","",Sequence)) %>% # isolate flanks
  select(flank = Sequence, estimate, std.error) %>%
  mutate(kd = estimate/1E9,
         stdE.err = (std.error/estimate)*0.593,
         stdE = log(kd) * 0.593,
         target = flank) %>%  
  inner_join(.,dfVar) %>%
  group_by(flank) %>%
  mutate(addVals = diMean(flank,dfVar)) %>%
  ungroup() %>%
  mutate(addVals = addVals/max(addVals),
         diVals = addVals + coef) # normalize

# benchmark plot
lm(stdE~addVals,stdEpi) %>% glance() # mononucleotide 
lm(stdE~diVals,stdEpi) %>% glance() # dinucleotide 

S13 = ggplot(stdEpi, aes(stdE,addVals)) + 
  presentation + 
  geom_errorbarh(aes(x = stdE, xmax = stdE + stdE.err, xmin = stdE -stdE.err,color = abs(coef)), size = 0.25) +
  geom_point(aes(color = abs(coef)), size = 0.01) +
  theme(aspect.ratio = 1,
        legend.key.size = unit(0.2,"cm")) +
  ylab("Relative sum of LASSO\nmononucleotide coefficients") +
  xlab(expression(bold(paste("Titration ", Delta, "G, kcal/mol")))) +
  scale_color_gradient("LASSO dinucleotide\ncoefficient magnitude",high = "red", low = "black")

ggsave("~/Flanks/Images/S13.png", plot = S13, height = 4, width = 10, units = "cm")
```

## Figure S14: NNet model(ensemble) vs. Pho4 titration data
```{r}
# join NN results with titration data
nnCor = 
  left_join(std_kd %>% select(flank,stdE,stdE.err,protein), countsNN.df%>% select(flank,ddG,protein))

# lm to extract linear equation
slopeDf = foreach(i = c("pho4","cbf1"), .combine = "rbind")%do%{
  iSlice = nnCor %>% filter(protein == i)
  mod = lm(stdE ~ ddG, iSlice)
  tidy(mod) %>% 
    mutate(term = ifelse(term == "(Intercept)","int","slope")) %>%
    select(term,estimate) %>%
    spread(term,estimate) %>%
    mutate(protein = i)
}

nnfitCor = 
  left_join(nnCor,slopeDf) %>%
  mutate(protein = ifelse(protein == "pho4","Pho4","Cbf1"), 
         protein = factor(protein, levels = c("Pho4","Cbf1")),
         ddG = (ddG*slope) + int) %>%
  na.omit()

S14 = ggplot(nnfitCor,aes(stdE,ddG)) +
  presentation +
  geom_errorbarh(aes(x = stdE, xmax = stdE + stdE.err, xmin = stdE - stdE.err), size = 0.2) +
  geom_point(size = 0.01) +
  theme(aspect.ratio = 1) +
  ylab(expression(bold(paste("Sequencing ", Delta, "G, kcal/mol")))) +
  xlab(expression(bold(paste("Titration ", Delta, "G, kcal/mol")))) +
  facet_grid(. ~ protein, scales = "free")

ggsave("~/Flanks/Images/S14.png", plot = S14, height = 4, width = 6, units = "cm")
```


##Table S8: titration binding sequences
```{r}
fread("~/Flank_seq/data/titrationseq_di.txt", header = F) %>% 
  rename(Index = V1, Sequence = V2) %>% 
  mutate(Sequence = gsub("CAATACACTGTTATC ","",Sequence),
         Sequence = gsub(" CTACTCGTTCGGTTATCCGGCGGTATGAC","",Sequence)) %>% 
  select(Sequence) %>% 
  xtable() %>% 
  print()
```


## Table S6: epistasis binding results
```{r}
fread("~/Flank_seq/data/stdEpistasis1.csv", header = T) %>%
  tbl_df() %>%
  mutate(Sequence = gsub(" ","",Sequence)) %>% # remove white space
  mutate(Sequence = substring(Sequence,16,31), Sequence = gsub("CACGTG","_",Sequence)) %>% # isolate flanks
  select(flank = Sequence, estimate, std.error) %>%
  mutate(kd = estimate/1E9,
         stdE.err = (std.error/estimate)*0.593,
         stdE = log(kd) * 0.593,
         target = flank) %>% 
  select(Sequence = flank, estimate, std.error) %>% 
  mutate(estimate = round(estimate), std.error = round(std.error)) %>% 
  arrange(estimate) %>% 
  bind_rows(.,pho4EpiCurves %>% 
              group_by(flank) %>% 
              summarise(estimate = mean(estimate), std.error = mean(std.error)) %>% 
              filter(flank == "negative control") %>% 
              rename(Sequence = flank)) %>% 
  xtable(digits = 0) %>% 
  print()   
```


## Figure S14: Cbf1 + Pho4 LASSO gapped position distribution
```{r}
dfSpace2 = 
  dfSpace %>%
  group_by(protein) %>%
  mutate(Observed = count/sum(count)) %>%
  select(protein,Observed,space) %>%
  spread(protein,Observed) %>%
  mutate(correct = c(8,8:1), 
         Theoretical = correct/sum(correct),
         Pho4 = Pho4/sum(Pho4),
         Cbf1 = Cbf1/sum(Cbf1)) %>%
  gather(protein,value,Pho4,Cbf1) %>%
  mutate(protein = factor(protein, levels = c("Pho4","Cbf1"))) %>% 
  gather(type,density,Theoretical,value)

#plot
colPal = brewer.pal(9,"Set1")
S14 = ggplot(dfSpace2) +
  presentation +
  geom_bar(aes(space,density, fill = type), stat = "identity", position = "dodge") +
  theme(aspect.ratio = 1,
        legend.key.size = unit(0.2, "cm")) +
  ylab("Probability density") +
  xlab("Number of gapped positions") +
  facet_grid(protein ~ .) +
  scale_fill_manual("", labels = c("Theoretical","Observed"), values = colPal[1:2])

ggsave("~/Flanks/Images/S14.png", plot = S14, height = 7, width = 7, units = "cm")
```

## Panel D: Composite additive of Pho4 + Cbf1 correlated to titration data (scaled and offset)
```{r}
# per-sequence correlation
allseqDat = 
  counts.df %>% 
  inner_join(std_kd,.,by=c("flank","protein")) %>%
  select(flank,stdE,stdE.err,ddG,protein)

# additive model correlation (extracting fit error too)
alladdDat = foreach(i = 1:2, .combine = "rbind")%do%{
  protSet = c("pho4","cbf1")
  tmp = counts.df %>% filter(protein == protSet[i]) %>% select(protein, flank, X01:X10,ddG)
  mod = lm(ddG ~ X01 + X02 + X03 + X04 + X05 + X06 + X07 + X08 + X09 + X10 , data = tmp)
  predMod = predict(mod, std_kd %>% filter(protein == protSet[i]))
  left_join(std_kd %>% filter(protein == protSet[i]) %>% select(protein,stdE,stdE.err,flank) %>% mutate(.fitted = predMod), tmp %>% select(protein,flank,ddG))
}

# combine per-sequence and additive model data
crossallDat = 
  alladdDat %>% 
  gather(key,value,ddG,.fitted) %>%
  distinct() %>%
  mutate(key = ifelse(key == "ddG","Individual ddG","Mononucleotide model"), 
         protein = ifelse(protein == "pho4","Pho4","Cbf1"),
         protein = factor(protein, levels = c("Pho4","Cbf1")))

# all Pearson r2
crossallDat %>%
  na.omit() %>%
  group_by(protein,key) %>%
  summarise(r2 = cor(stdE,value)^2) %>%
  kable()

# lm of titration vs ddG to extract slope
slopeDf = foreach(i = c("Pho4","Cbf1"), .combine = "rbind")%do%{
  iSlice = crossallDat %>% filter(protein == i, key == "Mononucleotide model") %>% distinct()
  mod = lm(stdE ~ value, iSlice)
  tidy(mod) %>% 
    mutate(term = ifelse(term == "(Intercept)","int","slope")) %>%
    select(term,estimate) %>%
    spread(term,estimate) %>%
    mutate(protein = i)
}

# scale seq data by titration data
crossallDat2 = 
  crossallDat %>%
  left_join(.,slopeDf) %>%
  filter(key == "Mononucleotide model") %>%
  mutate(key = "mononucleotide model",
         value = (value*slope) + int, 
         protein = factor(protein, levels = c("Pho4","Cbf1")))

# plot
p1 = ggplot(crossallDat2, aes(stdE,value)) +
  presentation +
  geom_abline(slope = 1, intercept = 0, lty = 2, color = "red") + # centered on middle of x-scale
  geom_errorbarh(aes(x = stdE, xmax = stdE + stdE.err, xmin = stdE - stdE.err),size = 0.25) +
  geom_point(size = 0.25) +
  facet_grid(key ~ protein, scale = "free") +
  theme(aspect.ratio = 1,
        axis.text.x = element_text(angle = 45, vjust = 0.5)) +
  ylab(expression(bold(paste("Sequencing ", Delta, "G, kcal/mol")))) +
  xlab(expression(bold(paste("Titration ", Delta, "G, kcal/mol")))) 

ggsave("~/Flanks/Images/f3pD.svg", plot = p1, height = 4.5, width = 8.7, units = "cm")

# Analysis
# output all data scaled and offset
allScaled = foreach(i = 1:2, .combine = "rbind")%do%{
  protSet = c("pho4","cbf1")
  tmp = counts.df %>% filter(protein == protSet[i]) %>% select(protein, flank, X01:X10,ddG)
  mod = lm(ddG ~ X01 + X02 + X03 + X04 + X05 + X06 + X07 + X08 + X09 + X10 , data = tmp)
  predMod = predict(mod, tmp %>% select(X01:X10) %>% distinct())
  tmp %>%
    select(X01:X10) %>%
    distinct() %>%
    mutate(.fitted = predMod, protein = i)
}

# return Kd range
allScaled %>%
  tbl_df() %>%
  mutate(protein = ifelse(protein == 1, "Pho4","Cbf1")) %>%
  left_join(.,slopeDf) %>%
  mutate(.scaled = (.fitted*slope) + int,
         Kd = exp(.scaled/0.593)*1E9) %>%
  group_by(protein) %>%
  summarise(minval = min(Kd), maxval = max(Kd), medval = median(Kd))

# return dG range
allScaled %>%
  tbl_df() %>%
  mutate(protein = ifelse(protein == 1, "Pho4","Cbf1")) %>%
  left_join(.,slopeDf) %>%
  mutate(.scaled = (.fitted*slope) + int,
         Kd = exp(.scaled/0.593)*1E9) %>%
  group_by(protein) %>%
  summarise(minval = min(.scaled), maxval = max(.scaled), medval = median(.scaled))
```


# Figure S7: Pho4 read consolidation
```{r}
### 10-mer vs 8-mer proof
allseqDat = 
  counts.df %>% 
  filter(protein == "pho4") %>%
  inner_join(std_kd %>% filter(protein == "pho4"),.) %>%
  select(flank,stdE,stdE.err,ddG)

alladdDat = 
  lm(ddG ~ X02 + X03 + X04 + X05 + X06 + X07 + X08 + X09 , data = counts.df %>% filter(protein == "pho4") %>% select(X01:X10,ddG)) %>%
  predict(.,std_kd %>% filter(protein == "pho4") %>% select(X01:X10)) %>%
  mutate(std_kd %>% filter(protein == "pho4") %>% select(flank,stdE), .fitted = .) 

crossallDat = 
  left_join(allseqDat,alladdDat) %>%
  gather(key,value,ddG,.fitted) %>%
  mutate(key = ifelse(key == "ddG","Individual ddG","Mononucleotide model"))

cor(alladdDat$stdE,alladdDat$.fitted)^2
###

# Collapse the most distal nucleotides to yield 8-mer data
read_data_8mer = 
  counts.df %>% 
  filter(protein == "pho4") %>%
  mutate(flank = substr(flank, 2, 9)) %>% 
  select(flank:X10, protein, rep)

# Combine all species with the same 8-mer core
## 8-mer seq data
seq8mer <- read_data_8mer %>% 
  group_by(flank, protein, rep) %>% 
  summarize(tf_count = sum(tf_count),
            ref_count = sum(ref_count)) %>%
  group_by(protein,rep) %>%
  mutate(tf_count = tf_count/sum(tf_count), # normalize to depth
         ref_count = ref_count/sum(ref_count),
         ddG = -log(tf_count/ref_count) * 0.593) %>%
  separate(flank,paste0("X0",1:8),1:7) %>%
  ungroup()

## 8-mer titration data
std8mer = 
  std_kd %>%
  filter(protein == "pho4") %>%
  select(flank,stdE,stdE.err,protein,estimate) %>%
  mutate(flank = substr(flank, 2, 9)) %>% 
  group_by(flank, protein) %>% 
  mutate(stdE = mean(stdE), stdE.err = mean(stdE.err), target = flank) %>%
  separate(target,paste0("X0",1:8),1:7) %>%
  ungroup()

# 8-mer additive model
add8mer = foreach(i = 1:4, .combine = "rbind")%do%{
  tmp = seq8mer %>% filter(rep == i) %>% select(X01:X08,ddG)
  lm(ddG ~ X01 + X02 + X03 + X04 + X05 + X06 + X07 + X08, data = tmp) %>%
  predict(.,std8mer %>% select(X01:X08)) %>%
  mutate(std8mer %>% select(flank,stdE), .fitted = ., rep = i)
} %>%
  group_by(rep) %>%
  summarise(Add_r2 = cor(stdE,.fitted)^2)

# 8-mer depth and species
depth8mer = 
  read_data_8mer %>%
  group_by(flank, rep) %>% 
  summarize(tf_count = sum(tf_count),
            ref_count = sum(ref_count)) %>%
  group_by(rep) %>% 
  summarize(tf_count = sum(tf_count),
            ref_count = sum(ref_count),
            depth = min(c(tf_count,ref_count))/n()) %>%
  select(rep,depth)

# 8-mer aggregate Pearson correlation test
cor8mer = 
  left_join(std8mer,seq8mer) %>% 
  select(stdE, stdE.err, ddG,rep) %>%
  group_by(rep) %>%
  summarise(r2 = cor(stdE, ddG)^2) %>%
  left_join(.,add8mer) %>%
  left_join(.,depth8mer) %>%
  mutate(length = 8)

# 10-mer seq data
seq10mer = 
  counts.df %>%
  filter(protein == "pho4")

# 10-mer additive data
add10mer = foreach(i = 1:4, .combine = "rbind")%do%{
  tmp = seq10mer %>% filter(rep == i) %>% select(X01:X10,ddG)
  lm(ddG ~ X01 + X02 + X03 + X04 + X05 + X06 + X07 + X08 + X09 + X10, data = tmp) %>%
  predict(.,std_kd %>% filter(protein == "pho4") %>% select(X01:X10)) %>%
  mutate(std_kd %>% filter(protein == "pho4") %>% select(flank,stdE), .fitted = ., rep = i)
} %>%
  group_by(rep) %>%
  summarise(Add_r2 = cor(stdE,.fitted)^2)  

# 8-mer depth and species
depth10mer = 
  counts.df %>%
  filter(protein == "pho4") %>%
  group_by(flank, rep) %>% 
  summarize(tf_count = sum(tf_count),
            ref_count = sum(ref_count)) %>%
  group_by(rep) %>% 
  summarize(tf_count = sum(tf_count),
            ref_count = sum(ref_count),
            depth = min(c(tf_count,ref_count))/n()) %>%
  select(rep,depth)

# 10-mer aggregate Pearson correlation test
cor10mer = 
  left_join(std_kd %>% filter(protein == "pho4") %>% select(flank,stdE,stdE.err),seq10mer) %>% 
  select(stdE, stdE.err, ddG,rep) %>%
  group_by(rep) %>%
  summarise(r2 = cor(stdE, ddG)^2) %>%
  left_join(.,add10mer) %>%
  left_join(.,depth10mer) %>%
  mutate(length = 10)

# Aggregate correlations
corDepth = 
  bind_rows(cor8mer,cor10mer) %>%
  rename(Individual = r2, Additive = Add_r2, Mean_counts = depth, Length = length, Replicate = rep) %>%
  gather(Pearson_cor,value,Individual,Additive)

# plot v2
corDepth2 = 
  bind_rows(cor8mer,cor10mer) %>%
  rename(Individual = r2, Additive = Add_r2, Mean_counts = depth, Length = length, Replicate = rep) %>%
  gather(Pearson_cor,value,Individual,Additive) %>%
  mutate(Length = ifelse(Length == 8,"16x","1x"), 
         Length = factor(Length, levels = c("1x","16x")),
         Pearson_cor = factor(Pearson_cor, 
                              levels = c("Individual","Additive")))

colPal = brewer.pal(9,"Set1")
S7 = ggplot(corDepth2) +
  presentation +
  geom_bar(aes(as.factor(Length),value,fill=Pearson_cor,group=Pearson_cor),stat = "summary", fun.y = "mean",position = position_dodge(width = 0.5), width = 0.5) +
  geom_point(aes(as.factor(Length),value,group=Pearson_cor),position = position_dodge(width = 0.5),size = 0.01) +
  theme(aspect.ratio = 1,
        legend.key.size = unit(0.3, "cm")) +
  ylab(expression(bold(paste("Pearson r"^"2")))) +
  xlab(expression(bold(paste("Mean reads per sequence, fold-change")))) +
  ylim(c(0,1)) +
  scale_fill_manual("", 
                     values = colPal[1:2],
                     labels = expression(bold(paste("individual ",Delta,Delta,"G")),
                                         bold("mononucleotide model")),
                     guide = guide_legend(label.vjust = 0.5,
                                          label.hjust = 0))

ggsave("~/Flanks/Images/S7.png", plot = S7, height = 5, width = 10, units = "cm")
```
