---
title: "Final figures"
author: "Dan"
date: "8/15/2017"
output: html_document
---

```{r setup, include=FALSE}
source("~/Flank_seq/Figures/setup.R")

presentation <-  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5,
                                  family = "Arial",
                                  face = "bold",
                                  size = 12),
        axis.title = element_text(family = "Arial",
                                  face = "bold",
                                  size = 6),
        axis.text = element_text(family = "Arial",
                                 face = "bold",
                                 size = 6),
        legend.text = element_text(family = "Arial",
                                   face = "bold",
                                   size = 6),
        legend.title = element_text(family = "Arial",
                                    face = "bold",
                                    size = 6),
        strip.text = element_text(family = "Arial",
                                  face = "bold",
                                  size = 6))
```

#Figure 1:
## Panel A: DNA design
## Panel B: Device operation
## Panel C: Sequencing quantitation
## Panel B: Landscape
```{r}

# foreach(prot = c("pho4","cbf1"))%dopar%{
#   bestFlank = 
#     countsNN.df %>% 
#     filter(protein == prot) %>% 
#     arrange(ddG) %>%
#     slice(1) %>%
#     .$flank
#   
#   shockDf = 
#     countsNN.df %>% 
#     filter(protein == prot) %>% 
#     select(flank,ddG) %>%
#     mutate(cc = hdset2(flank,bestFlank)) %>% 
#     group_by(cc) %>% 
#     arrange(flank) %>%
#     mutate(t = seq(0,360-(360/n()),length.out = n()) * pi / 180) %>%
#     ungroup()
#   
#   write.table(shockDf, paste0("~/Flank_seq/data/shock_",prot,".txt"), col.names = T, row.names = F, quote = F, sep = "\t")
# }

shockDf = 
  bind_rows(fread(paste0("~/Flank_seq/data/shock_pho4.txt"), header = T) %>% mutate(protein = "pho4"),
            fread(paste0("~/Flank_seq/data/shock_cbf1.txt"), header = T) %>% mutate(protein = "cbf1")) %>%
  tbl_df()

# join NN results with titration data
nnCor = 
  left_join(std_kd %>% select(flank,stdE,protein), shockDf %>% select(flank,ddG,cc,t))

# lm to extract linear equation
slopeDf = foreach(i = c("pho4","cbf1"), .combine = "rbind")%do%{
  iSlice = nnCor %>% filter(protein == i)
  mod = lm(stdE ~ ddG, iSlice)
  tidy(mod) %>% 
    mutate(term = ifelse(term == "(Intercept)","int","slope")) %>%
    select(term,estimate) %>%
    spread(term,estimate) %>%
    mutate(protein = i)
}

nnfitCor = 
  left_join(shockDf,slopeDf) %>%
  mutate(ddG = (ddG*slope), t = t/2) %>% #only scale, no offset, create hemisphere
  select(-int,-slope) %>%
  group_by(cc) %>% 
  mutate(meanval = mean(ddG)) %>%
  ungroup() %>%
  bind_rows(data.frame(flank = NA, ddG = NA, cc = 11, t = 2*pi, protein = "pho4", int = NA, slope = NA))
  
labelDf = 
  nnfitCor %>%
  filter(protein == "pho4") %>%
  group_by(cc,protein) %>%
  summarise(count = n()) %>%
  ungroup() %>%
  mutate(rolling = cumsum(count), 
         digits = nchar(as.character(count)), 
         filling = paste0(count, "   "))

###
prot = "pho4"
###

# circle plot
p1 = ggplot(nnfitCor %>% filter(protein == prot)) +
  geom_point(aes(t,cc,color=ddG)) +
  coord_polar(start = -pi/2) +
  ylim(c(0,10)) +
  scale_color_gradient2("ddG, kcal/mol",low = "#0057E7", high = "#D62D20", mid = "white") +
  theme(aspect.ratio = 1, 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none",
        legend.text = element_text(angle = 45)) +
  geom_text(data = labelDf %>% filter(cc != 0, cc != 11), aes(pi,cc,label =cc), inherit.aes = F, vjust = 1.5) +
  geom_text(data = labelDf %>% filter(cc != 11), aes(0,cc,label =filling), inherit.aes = F, angle = 45, hjust= 1, color = "gray")

ggsave("~/Flanks/Images/f1pD.png", plot = p1, height = 12, width = 12, units = "cm")
```

# Figure 2:
## Panel A: Monte Carlo sampling of Spec-seq PWM and ddG

```{r}
### Parameters ###
iVal = unique(c(seq(100,1000,100),seq(1E3,1E4,1E3),seq(1E4,1E5,1E4),seq(1E5,1E6,1E5),seq(1E6,1E7,1E6)))
jVal = 1:10

testCount = lapply(list(iVal,jVal),function (x) x %>% length()) %>%
  unlist() %>%
  prod()

typeCor = "spearman"
###

# read-in data
readDat2 =
  fread("~/Flank_seq/data/Stormo_All_Expts.txt", header = T) %>%
  tbl_df() %>%
  rename(Seq = Sequence, True_Bound = Bound, True_Unbound = Unbound, True_Ratio = Bound.Unbound) %>%
  mutate(index = c(rep(1,256), rep(2,256), rep(3,1024), rep(4,1024))) %>% #R2,R4,R3.1,R3.2 (only from Expt 1)
  filter(index == 4) %>%
  mutate(True_Bound_P = True_Bound / sum(True_Bound),
         True_Unbound_P = True_Unbound / sum(True_Unbound),
         spec_ddG = True_Bound_P / True_Unbound_P, target = Seq) %>%
    separate(target, paste0("Var",1:10),1:9) %>%
  select(Seq,spec_ddG,Var5,Var7:Var10)

# extract PWM coefficients from real data
refCoef =
  lm(spec_ddG ~ Var5+Var7+Var8+Var9+Var10, data = readDat2) %>%
  coef() %>%
  as.vector()

# perform sampling
dsPWM = foreach(i = iVal, .combine = "rbind")%dopar%{ # sampling depth split equally between bound and unbound
  libDepth = i
  dnaSpecies = 1024
  foreach(j = jVal, .combine = "rbind")%do%{ # iterations

    trueRatioDf =
      data.frame(specIndex = readDat2$Seq, trueRatio = readDat2$spec_ddG) %>%
      tbl_df()

    obsRatioDf =
      inner_join(tbl_df(data.frame(sampleBound = sample(trueRatioDf$specIndex, libDepth/2, prob = trueRatioDf$trueRatio, replace = T))) %>%
        count(sampleBound) %>%
        rename(specIndex = sampleBound, bound = n),
        tbl_df(data.frame(sampleUnbound = sample(trueRatioDf$specIndex, libDepth/2, prob = rep(1,dnaSpecies), replace = T))) %>%
        count(sampleUnbound) %>%
        rename(specIndex = sampleUnbound, unbound = n)) %>%
      mutate(obsRatio = bound/unbound)

    ratioDf =
      inner_join(trueRatioDf, obsRatioDf %>% select(-bound, -unbound)) %>%
      mutate(target = specIndex) %>%
      separate(target, paste0("Var",1:10),1:9) %>%
      select(-Var1,-Var2,-Var3,-Var4,-Var6)

    iterCoef = tryCatch({lm(obsRatio ~ Var5+Var7+Var8+Var9+Var10, data = ratioDf) %>%
        coef() %>%
        as.vector()}, error = function(err) {NA})

    rSeq = tryCatch({ifelse(cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$p.value < (0.05/testCount),cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$estimate,NA)}, error = function(err) {NA})
    rPwm = tryCatch({ifelse(cor.test(refCoef, iterCoef, method = typeCor)$p.value < (0.05/testCount),cor.test(refCoef, iterCoef, method = typeCor)$estimate,NA)}, error = function(err) {NA})

    data.frame(readTotal = libDepth, meanRead_perSeq = (libDepth/2)/dnaSpecies, iter = j, r2seq = ifelse(is.na(rSeq),0,rSeq), r2pwm = ifelse(is.na(rPwm),0,rPwm), libSize = dnaSpecies, seqObsPct = nrow(obsRatioDf)/dnaSpecies) %>%
      mutate(r2seq = ifelse(typeCor == "pearson", r2seq^2, r2seq), r2pwm = ifelse(typeCor == "pearson", r2pwm^2, r2pwm))
  }
} %>% gather(key,value,c(r2seq,r2pwm))

write.table(dsPWM, "~/Flank_seq/data/dspwmSimulation.txt", col.names = T, row.names = F, quote = F)
###

dsPWM = fread("~/Flank_seq/data/dspwmSimulation.txt", header = T)

#plot
colPal = brewer.pal(9,"Set1")
p1 = ggplot(dsPWM) +
  presentation +
  ylim(c(0,1)) +
  geom_boxplot(aes(readTotal,value,group = interaction(readTotal,key),color=key), outlier.alpha = 0, position = "dodge", width = 0.05) +
  geom_line(aes(readTotal,value,color = key), stat="summary",fun.y = "median", size = 0.5) +
  scale_x_log10() +
  scale_color_manual(label = c("Additive model","Individual ddG"),"", values = colPal[1:2]) +
  theme(aspect.ratio = 1/2) +
  ylab("Accuracy\n(Spearman's rho)") +
  xlab("Sequencing depth, read counts")

ggsave("~/Flanks/Images/f2pA.eps", plot = p1, height = 3, width = 8.7, units = "cm")
```

## Panel B: Simulate assay output as function of defined parameters.

Assumptions: 
1) uniform sampling of reads
2) uniform sequence distribution
3) symmetric energy range
4) energies normally distributed around dG = 0 with stdev = energy range/6 (~99% of range)

Arg:
dnaSpecies = number of possible DNA sequences
readNum = depth of sequencing (sum of bound and unbound sublibs)
gibbsDiff = the free energy difference between strongest and weakest binders

Output:
dataframe of paramters and accuracy of observed ratios as measured by r^2 compared to true values


```{r, eval=FALSE}
### Parameters ###
# mVal = c(0.25,0.5,1:5)
# nVal = 10^(2:6)
# iVal = 10^(3:8)  
# jVal = 1:10
# 
# testCount = lapply(list(mVal,nVal,iVal,jVal),function (x) x %>% length()) %>%
#   unlist() %>%
#   prod()
# 
# typeCor = "spearman"
# ###
# 
# returnDf =
#   foreach(m = mVal, .combine = "rbind")%do%{ # ddG = 0.25-5.0 corresponds to 1.5- to 4590- fold change
#     foreach(n = nVal, .combine = "rbind")%do%{ # library size
#       gibbsDiff = m
#       dnaSpecies = n
#       dnaRatios = sapply(rnorm(dnaSpecies, mean = 0, sd = gibbsDiff/6), function(x) exp(-x/.593)) #RT at 298K
#       foreach(i = iVal, .combine = "rbind")%do%{ # sampling depth
#         libDepth = i
#         print(paste(m,n,i))
#         foreach(j = jVal, .combine = "rbind")%dopar%{ # 10 iterations
#           trueRatioDf = 
#             tbl_df(data.frame(specIndex = 1:dnaSpecies, trueRatio = dnaRatios)) %>%
#             mutate(prob = trueRatio/sum(trueRatio))
#           obsRatioDf =
#             inner_join(tbl_df(data.frame(sampleBound = sample(trueRatioDf$specIndex, libDepth/2, prob = trueRatioDf$prob, replace = T))) %>%
#               count(sampleBound) %>%
#               rename(specIndex = sampleBound, bound = n),
#               tbl_df(data.frame(sampleUnbound = sample(trueRatioDf$specIndex, libDepth/2, prob = rep(1/dnaSpecies,dnaSpecies), replace = T))) %>%
#               count(sampleUnbound) %>%
#               rename(specIndex = sampleUnbound, unbound = n)) %>% 
#             mutate(obsRatio = bound/unbound)
#           ratioDf = 
#             inner_join(trueRatioDf %>% select(-prob), obsRatioDf %>% select(-bound, -unbound))
#           rSqd = 
#             tryCatch({ifelse(cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$p.value < (0.05/testCount), cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$estimate, NA)}, error = function(err) {NA})
#           
#           data.frame(readTotal = libDepth, meanRead_perSeq = (libDepth/2)/dnaSpecies, iter = j, r2 = ifelse(is.na(rSqd),0,rSqd), ddG = gibbsDiff, libSize = dnaSpecies, seqObsPct = nrow(obsRatioDf)/dnaSpecies) %>% 
#             mutate(r2 = ifelse(typeCor == "pearson", r2^2, r2), infoCov = r2 * seqObsPct)
#         }
#       }
#     }
#   }
# 
# write.table(returnDf, "~/Flank_seq/data/modelSimulation.txt", col.names = T, row.names = F, quote = F)
###

returnDf2 =
  fread("~/Flank_seq/data/modelSimulation.txt", header = T) %>% 
  gather(key,value,r2,seqObsPct) %>% 
  tbl_df() %>%
  mutate(key = ifelse(key == "r2", "Spearman's rho", "Fract. Obs."),
         key = factor(key, levels = c("Spearman's rho", "Fract. Obs.")))

#plot

colPal = colorRampPalette(brewer.pal(9,"YlOrRd"))(10)
p1 = ggplot(returnDf2 ) +
  presentation +
  geom_line(aes(ddG, value, color = as.factor(readTotal), group = as.factor(readTotal)),stat = "summary",fun.y = "median", na.rm = T, size = 0.5) +
  geom_point(aes(ddG, value, color = as.factor(readTotal), group = as.factor(readTotal)),stat = "summary",fun.y = "median", na.rm = T, size = 0.5) +
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  facet_grid(libSize ~ key, scales = "free") +
  ylim(c(0,1)) +
  xlab("ddG, kcal/mol") +
  ylab("") +
  scale_color_manual("Sequencing depth", values = colPal[(length(colPal)-6):length(colPal)])

ggsave("~/Flanks/Images/f2pB.eps", plot = p1, height = 10.5, width = 8.7, units = "cm")
```

## Panel C: Flank library depth simultion
```{r}
# FYI spec-seq = 5.4kt (3.2 kcal/mol) affinity range
# readDat = 
#   fread("~/Flank_seq/data/Stormo_All_Expts.txt", header = T) %>%
#   tbl_df() %>%
#   rename(Seq = Sequence, True_Bound = Bound, True_Unbound = Unbound, True_Ratio = Bound.Unbound) %>%
#   mutate(index = c(rep(1,256), rep(2,256), rep(3,1024), rep(4,1024))) %>% #R2,R4,R3.1,R3.2 (only from Expt 1)
#   group_by(index) %>%
#   mutate(True_Bound_P = True_Bound / sum(True_Bound),
#          True_Unbound_P = True_Unbound / sum(True_Unbound),
#          spec_ddG = True_Bound_P / True_Unbound_P) %>%
#   ungroup() %>%
#   mutate(Seq = 1:nrow(.)) %>%
#   select(Seq,spec_ddG) %>% 
#   right_join(.,data.frame(Seq = 1:(4^10)) %>%
#   mutate(ref_ddG = sapply(rnorm(4^10, mean = 0, sd = 1/6), function(x) exp(-x/.593)))) 
# 
# ### Parameters ###
# iVal = unique(c(seq(1E1,1E2,1E1),seq(1E2,1E3,1E2),seq(1E3,1E4,1E3),seq(1E4,1E5,1E4),seq(1E5,1E6,1E5),seq(1E6,1E7,1E6),seq(1E7,1E8,1E7),seq(1E8,1E9,1E8)))
# jVal = 1:10
# kVal = c("spec_ddG","ref_ddG")
# 
# testCount = lapply(list(iVal,jVal,kVal),function (x) x %>% length()) %>%
#   unlist() %>%
#   prod()
# ###
# 
# dsDat = foreach(i =iVal, .combine = "rbind")%do%{ # sampling depth split equally between bound and unbound
#         print(i)
#         foreach(k = kVal, .combine = "rbind")%do%{ # spec-seq or theoretical normal
#           foreach(j = jVal, .combine = "rbind")%dopar%{ # iterations
#             
#             libDepth = i
#             
#             dnaSpecies = length(readDat[,k] %>% unlist(use.names=F) %>% na.omit())
#             
#             trueRatioDf = 
#               tbl_df(data.frame(specIndex = 1:dnaSpecies, trueRatio = readDat[,k] %>% unlist(use.names=F) %>% na.omit()))
#             
#             obsRatioDf =
#               inner_join(tbl_df(data.frame(sampleBound = sample(trueRatioDf$specIndex, libDepth/2, prob = trueRatioDf$trueRatio, replace = T))) %>%
#                 count(sampleBound) %>%
#                 rename(specIndex = sampleBound, bound = n),
#                 tbl_df(data.frame(sampleUnbound = sample(trueRatioDf$specIndex, libDepth/2, prob = rep(1,dnaSpecies), replace = T))) %>%
#                 count(sampleUnbound) %>%
#                 rename(specIndex = sampleUnbound, unbound = n)) %>% 
#               mutate(obsRatio = bound/unbound)
#             
#             ratioDf = 
#               inner_join(trueRatioDf, obsRatioDf %>% select(-bound, -unbound))
#             
#             rSqd = tryCatch({ifelse(cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$p.value < (0.05/testCount),cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$estimate,NA)}, error = function(err) {NA})
#             
#             error = tryCatch({ifelse(cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = "pearson" )$p.value < (0.05/testCount),cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = "pearson")$estimate^2,NA)}, error = function(err) {NA})
#             
#             data.frame(readTotal = libDepth, meanRead_perSeq = (libDepth/2)/dnaSpecies, iter = j, r2 = ifelse(is.na(rSqd),0,rSqd), err = ifelse(is.na(error),1,1-error),libSize = dnaSpecies, seqObsPct = nrow(obsRatioDf)/dnaSpecies) %>% 
#               mutate(type = k, r2 = ifelse(typeCor == "pearson", r2^2, r2))
#           }
#         }
# } %>% tbl_df()
# 
# write.table(dsDat, "~/Flank_seq/data/downsampleSimulation.txt", col.names = T, row.names = F, quote = F)
###

dsDat = fread("~/Flank_seq/data/downsampleSimulation.txt", header = T) %>% 
  filter(type == "ref_ddG") %>% 
  gather(key,value,r2,seqObsPct,err) %>%
  group_by(meanRead_perSeq,key) %>%
  summarise(medVal = median(value, na.rm = T))

#plot
colPal = brewer.pal(9,"Set1")
p1 = ggplot(dsDat) +
  presentation +
  geom_line(aes(meanRead_perSeq, medVal, color = key), size = 0.5) +
  geom_point(aes(meanRead_perSeq, medVal, color = key), size = 0.5) +
  theme(aspect.ratio = 1/2, 
        axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  xlab("Mean reads per sequence") +
  ylab("Median") +
  scale_x_log10(limits = c(0.1,1000)) +
  scale_color_manual(labels = c("Fract. Error","Spearman's rho","Fract. Observed"),"", values = colPal[1:3])

ggsave("~/Flanks/Images/f2pC.eps", plot = p1, height = 3, width = 8.7, units = "cm")

# predictions
dsDat %>%
  tbl_df() %>% 
  filter(key == "seqObsPct", medVal > 0.95) %>% 
  arrange(meanRead_perSeq)
```

# Figure 3:
## Panel A: Composite additive model of Pho4 and Cbf1
```{r}
# additive model plots
outAdd = addMotif(counts.df, "ddG") %>% 
  mutate(monuc.value = factor(monuc.value, levels = c("T","G","C","A"))) %>%  
  group_by(monuc.value,monuc.pos,protein,side) %>% 
  summarise(totalmean = mean(monuc.meanval), sem = sd(monuc.meanval)/sqrt(n())) %>% # SEM across replicates
  ungroup()
  
relAdd = foreach(i = c("pho4","cbf1"), .combine = "rbind")%do%{
  iSlice = outAdd %>% filter(protein == i)
  minRow = iSlice %>% filter(totalmean == min(totalmean))
  iSlice %>%
    mutate(affinitySEM = sqrt(((sem/totalmean)^2)+((minRow$sem/minRow$totalmean)^2)), 
           affinityFrac = totalmean/minRow$totalmean,
           affinitySEM = abs(affinityFrac*affinitySEM))
} %>% 
  mutate(protein = ifelse(protein == "pho4","Pho4","Cbf1"),
         protein = factor(protein, levels = c("Pho4","Cbf1"))) %>%
  left_join(., data.frame(monuc.value = c("A","C","G","T"), colPal = c("#55C341","#3036DF","#F0E328","#EF1717"))) %>% 
  mutate(colPal = as.character(colPal))
  
# composite textmap

p1 = ggplot(relAdd,aes(monuc.pos, affinityFrac)) +
  presentation +
  geom_errorbar(aes(x = monuc.pos, ymin = affinityFrac - affinitySEM, ymax = affinityFrac + affinitySEM),size = 0.2,width=0.2) +
  geom_text(aes(label = monuc.value, color = monuc.value), size = 2.5) +
  theme(aspect.ratio = 1, 
        legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  ylab("Relative affinity") +
  xlab("Flank position") +
  facet_grid(.~protein, scales = "free") +
  scale_color_manual(values = c("#55C341","#3036DF","#F0E328","#EF1717"))

ggsave("~/Flanks/Images/f3pA.eps", plot = p1, height = 4.5, width = 8.7, units = "cm")
```

## Panel B: PWM from literature of Pho4 and Cbf1
size = 3 x 8.7 cm
```{r}
pwmList = list("Pho4" = fread("~/Flank_seq/data/pwm_pho4.txt", header = F) %>% as.matrix(),
               "Cbf1" = fread("~/Flank_seq/data/pwm_cbf1.txt", header = F) %>% as.matrix())
row.names(pwmList[["Pho4"]]) = c("A","C","G","T")
row.names(pwmList[["Cbf1"]]) = c("A","C","G","T")

cs1 = make_col_scheme(chars=c('A', 'T', 'C', 'G'), 
                      groups=c('gr1', 'gr2', 'gr3', 'gr4'), 
                      cols=c("#55C341","#EF1717","#3036DF","#F0E328"))

# Generate sequence logo
p1 = ggplot() + 
  geom_logo(pwmList[["Pho4"]],col_scheme=cs1) +
  presentation +
  theme(legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        aspect.ratio = 1/2) +
  scale_x_continuous(breaks=1:8, labels=c("-1",rep("",6),"1"))

p2 = ggplot() +
  geom_logo(pwmList[["Cbf1"]],col_scheme=cs1) +
  presentation +
  theme(legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        aspect.ratio = 1/2) +
  scale_x_continuous(breaks=1:8, labels=c("-2","-1",rep("",6)))

pp = cowplot::plot_grid(p1,p2,nrow=1)

ggsave("~/Flanks/Images/f3pB.eps", plot = pp, height = 3, width = 8.7, units = "cm")

```

## Panel C: Selected binding curves
```{r}
pho4Curves = 
  fread("~/Flank_seq/data/pho4BindingCurves.csv", sep = ",", header = T) %>%
  tbl_df() %>%
  rename(ID = Onum, fBound = R, Conc = Ch) %>%
  mutate(index = paste("ID",ID,sep=""), index = factor(index, paste0("ID",1:32))) %>%
  createCurves(.) %>%
  mutate(index = factor(index, levels = paste0("ID",1:32))) %>%
  left_join(.,fread("~/Flank_seq/data/seq2ID_pho4.txt", header = F) %>%
    unite("flank",V3,V5,sep="_") %>%
    select(index = V1, flank) %>%
    mutate(index = paste("ID",index,sep=""))) %>%
  mutate(flank = ifelse(index == "ID32","neg ctrl",flank)) %>%
  mutate(flank = factor(flank, levels = arrange(.,estimate) %>% .$flank %>% unique()),
         estimate = percent_rank(estimate))

estSpread = pho4Curves %>% arrange(estimate) %>% .$estimate %>% unique()

#plot
colPal = brewer.pal(9,"Set1")
p1 = ggplot(pho4Curves %>% filter(estimate == estSpread[32]|estimate == estSpread[31]|estimate == estSpread[20]|estimate == estSpread[10]|estimate == estSpread[1])) +
  geom_point(aes(Conc,fBound,color=flank), size = 0.5) +
  geom_line(aes(Conc,.fitted,color=flank)) +
  presentation +
  theme(aspect.ratio = 1, 
        axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  ylab("Ratio DNA/protein") +
  xlab("Substrate concentration, nM") +
  xlim(c(0,100)) +
  scale_color_manual("", values = colPal)

ggsave("~/Flanks/Images/f3pC.eps", plot = p1, height = 4.5, width = 8.7, units = "cm")
```

## Panel D: Composite additive of Pho4 + Cbf1 correlated to titration data (scaled and offset)
```{r}
# per-sequence correlation
allseqDat = 
  counts.df %>% 
  inner_join(std_kd,.,by=c("flank","protein")) %>%
  select(flank,stdE,stdE.err,ddG,protein)

# additive model correlation (extracting fit error too)
alladdDat = foreach(i = 1:2, .combine = "rbind")%do%{
  protSet = c("pho4","cbf1")
  tmp = counts.df %>% filter(protein == protSet[i]) %>% select(protein, flank, X01:X10,ddG)
  mod = lm(ddG ~ X01 + X02 + X03 + X04 + X05 + X06 + X07 + X08 + X09 + X10 , data = tmp)
  predMod = predict(mod, std_kd %>% filter(protein == protSet[i]))
  left_join(std_kd %>% filter(protein == protSet[i]) %>% select(protein,stdE,stdE.err,flank) %>% mutate(.fitted = predMod), tmp %>% select(protein,flank,ddG))
}

# combine per-sequence and additive model data
crossallDat = 
  alladdDat %>% 
  gather(key,value,ddG,.fitted) %>%
  distinct() %>%
  mutate(key = ifelse(key == "ddG","Individual ddG","Mononucleotide model"), 
         protein = ifelse(protein == "pho4","Pho4","Cbf1"),
         protein = factor(protein, levels = c("Pho4","Cbf1")))

# all Pearson r2
crossallDat %>%
  na.omit() %>%
  group_by(protein,key) %>%
  summarise(r2 = cor(stdE,value)^2) %>%
  kable()

# lm of titration vs ddG to extract slope
slopeDf = foreach(i = c("Pho4","Cbf1"), .combine = "rbind")%do%{
  iSlice = crossallDat %>% filter(protein == i, key == "Mononucleotide model") %>% distinct()
  mod = lm(stdE ~ value, iSlice)
  tidy(mod) %>% 
    mutate(term = ifelse(term == "(Intercept)","int","slope")) %>%
    select(term,estimate) %>%
    spread(term,estimate) %>%
    mutate(protein = i)
}

# scale seq data by titration data
crossallDat2 = 
  crossallDat %>%
  left_join(.,slopeDf) %>%
  filter(key == "Mononucleotide model") %>%
  mutate(value = (value*slope) + int, 
         protein = factor(protein, levels = c("Pho4","Cbf1")))

# plot
p1 = ggplot(crossallDat2, aes(stdE,value)) +
  presentation +
  geom_abline(slope = 1, intercept = 0, lty = 2, color = "red") + # centered on middle of x-scale
  geom_errorbarh(aes(x = stdE, xmax = stdE + stdE.err, xmin = stdE - stdE.err),size = 0.25) +
  geom_point(size = 0.25) +
  facet_grid(key ~ protein, scale = "free") +
  theme(aspect.ratio = 1) +
  ylab("Sequencing dG, kcal/mol") +
  xlab("Titration dG, kcal/mol") 

ggsave("~/Flanks/Images/f3pD.eps", plot = p1, height = 4.5, width = 8.7, units = "cm")
write.table(crossallDat2,"~/Desktop/compositeSeq.txt", col.names = T, row.names = F, quote = F, sep = "\t")

# output all data scaled and offset
allScaled = foreach(i = 1:2, .combine = "rbind")%do%{
  protSet = c("pho4","cbf1")
  tmp = counts.df %>% filter(protein == protSet[i]) %>% select(protein, flank, X01:X10,ddG)
  mod = lm(ddG ~ X01 + X02 + X03 + X04 + X05 + X06 + X07 + X08 + X09 + X10 , data = tmp)
  predMod = predict(mod, tmp %>% select(X01:X10) %>% distinct())
  tmp %>% 
    select(X01:X10) %>% 
    distinct() %>%
    mutate(.fitted = predMod, protein = i)
}

# return Kd range
allScaled %>% 
  tbl_df() %>%
  mutate(protein = ifelse(protein == 1, "Pho4","Cbf1")) %>% 
  left_join(.,slopeDf) %>% 
  mutate(.scaled = (.fitted*slope) + int,
         Kd = exp(.scaled/0.593)*1E9) %>% 
  group_by(protein) %>% 
  summarise(minval = min(Kd), maxval = max(Kd), medval = median(Kd))

# return dG range
allScaled %>% 
  tbl_df() %>%
  mutate(protein = ifelse(protein == 1, "Pho4","Cbf1")) %>% 
  left_join(.,slopeDf) %>% 
  mutate(.scaled = (.fitted*slope) + int,
         Kd = exp(.scaled/0.593)*1E9) %>% 
  group_by(protein) %>% 
  summarise(minval = min(.scaled), maxval = max(.scaled), medval = median(.scaled))

```

# Figure 4:
## Panel A: B2 bomber of Pho4 + Cbf1

## Panel B: Pho4 + Cbf1 LASSO regression
```{r}
# Specify protein
protVal = "cbf1"

# Define number of interacting positions
intClass = 2

# Determine the number of terms at each model layer
modelTerms = 
  data.frame(intClass = 1:9) %>% 
  tbl_df() %>%
  mutate(intSet = 4^intClass, intPos = choose(10,intClass), intTerm = intSet * intPos)

# Determine positions of interaction
listDfPos = foreach(i = 1:intClass)%do%{
  expand.grid(rep(list(1:10),i)) %>%
  filter_(.dots = if(i > 1) {paste0("Var",2:i,">Var",1:(i-1))}) # remove redundancy
  }
  
###

# # Subset and encode dataframe for LASSO input 
# doMC::registerDoMC(cores = 40)
# 
# summDat = counts.df %>% filter(protein == protVal) %>% group_by(flank) %>% summarise(ddG = mean(ddG,na.rm = T))
# testDat = encodeTerms(summDat, 10000, 2, listDfPos)
# 
# # Classify predictor and response elements
# options(contrasts=c(ordered='contr.Dummy', unordered='contr.Dummy')) # no intercept
# x <- model.matrix( ~ .-1, testDat %>% select(contains("Var")))
# y <- testDat %>% .$ddG
# 
# # Perform LASSO regression
# fit.lasso=glmnet(x,y,alpha=1, intercept=FALSE, nlambda = 150) # no intercept, alpha = 1 for lasso, = 0 for ridge
# fitLasso = tidy(fit.lasso)
# write.table(fitLasso, paste0("~/Flank_seq/data/lassoFit_",protVal,".txt"), row.names = F, col.names = T, quote = F, sep = "\t")
# 
# # 10-fold cross validation
# cv.lasso=cv.glmnet(x,y, intercept=FALSE, parallel = T)
# cvLasso = tidy(cv.lasso) 
# write.table(cvLasso, paste0("~/Flank_seq/data/lassoCV_",protVal,".txt"), row.names = F, col.names = T, quote = F, sep = "\t")

# # Extract coefficients
# dfVar = 
#   coef(cv.lasso) %>% 
#   as.data.frame.matrix() %>% 
#   tbl_df() %>% 
#   mutate(vars = coef(cv.lasso) %>% rownames()) %>% 
#   rename(coef = `1`) %>% 
#   filter(vars != "(Intercept)") %>%
#   separate(vars, c("pos","term"),"[.]") %>%
#   group_by(coef,pos,term) %>%
#   mutate(row = substring(pos,4,nchar(pos)) %>% as.integer(),
#          flank = convertFlank(term,row,listDfPos)) %>%
#   ungroup() %>%
#   select(flank, coef) %>%
#   group_by(flank) %>%
#   mutate(space = (gregexpr("[A,C,G,T]", flank) %>% unlist() %>% diff() - 1) %>% sum(),
#          elements = gregexpr("[A,C,G,T]", flank) %>% unlist() %>% length()) %>%
#   ungroup()
# write.table(dfVar, paste0("~/Flank_seq/data/dfVar_",protVal,".txt"), row.names = F, col.names = T, quote = F, sep = "\t")

###

# LASSO plot
fitLasso = bind_rows(fread(paste0("~/Flank_seq/data/lassoFit_pho4.txt"), header = T) %>% mutate(protein = "pho4"),
                     fread(paste0("~/Flank_seq/data/lassoFit_cbf1.txt"), header = T) %>% mutate(protein = "cbf1")) %>% 
  separate(term,c("variable","base"),sep = "[.]") %>%
  mutate(dfrow = substring(variable, 4, nchar(variable)) %>% as.integer()) %>%
  group_by(estimate,lambda,dev.ratio,step) %>%
  mutate(flank = convertFlank(base,dfrow,listDfPos)) %>%
  ungroup()

lassoPlot = foreach(j = c("pho4","cbf1"), .combine = "rbind")%do%{
  jSlice = fitLasso %>% filter(protein == j)
  foreach(i = unique(jSlice %>% .$flank), .combine = "rbind")%do%{
    allLambda = jSlice %>% select(lambda) %>% distinct()
    iSlice = jSlice %>% select(lambda,estimate,flank,variable,base) %>% filter(flank == i)
    flankName = iSlice %>% .$flank %>% unique()
    varName = iSlice %>% .$variable %>% unique()
    baseName  = iSlice %>% .$base %>% unique()
    left_join(allLambda,iSlice %>% select(lambda,estimate), by = "lambda") %>%
      mutate(estimate = ifelse(is.na(estimate),0,estimate), 
             flank = flankName,
             flank = paste0(substr(flank,1,5),"_",substr(flank,6,10)),
             variable = varName, 
             base = baseName,
             protein = j,
             protein = ifelse(protein == "pho4","Pho4","Cbf1"),
             protein = factor(protein, levels = c("Pho4","Cbf1")))
  }
} %>% mutate(type = nchar(base),
             feature = ifelse(type == 1,"mononucleotide","dinucleotide"),
             feature = factor(feature, levels = c("mononucleotide","dinucleotide")))

labelDf2 = 
  lassoPlot %>%
  filter(estimate != 0) %>% 
  group_by(protein, feature, flank) %>%
  mutate(maxLamdha = max(lambda)) %>% 
  filter(maxLamdha == lambda) %>% 
  mutate(signGroup = sign(estimate)) %>% 
  group_by(protein, feature, signGroup) %>% 
  mutate(rank = dense_rank(lambda)) %>%
  filter(rank >= max(rank)-3) %>% 
  mutate(rank = dense_rank(lambda),
         id = row_number(),
         colVal = signGroup) %>%
  filter(id <= 3) %>% 
  ungroup() %>% 
  select(protein,flank,id,colVal,feature) %>% 
  left_join(.,data.frame(id = c(1:3,3:1), yPos = c(0.1,0.075,0.05,-0.1,-0.125,-0.15), colVal = c(1,1,1,-1,-1,-1)))

lassoPlot2 = 
  left_join(lassoPlot, labelDf2) %>% 
  mutate(colVal = ifelse(is.na(colVal),0,colVal))

p1 = ggplot(lassoPlot2) +
  presentation +
  geom_line(aes(lambda,estimate,group=interaction(variable,base), color = colVal), size = 0.5) +
  theme(aspect.ratio = 1, 
        legend.position = "none") +
  ylab("Variable coefficient") +
  xlab("Penalty coefficient, lambda") +
  scale_x_reverse() +
  geom_label(data = labelDf2,  
             aes(x = 0.055, y = yPos, label = flank, color = colVal), 
             inherit.aes = F, 
             size = 1.75, 
             label.padding = unit(0.1, "lines"),
             label.r = unit(0.05, "lines")) +
  scale_color_gradient2(high = "red", mid = "gray", low = "blue") +
  facet_wrap(~ protein+feature, nrow = 1)

ggsave("~/Flanks/Images/f4pB.eps", plot = p1, height = 8, width = 17.8, units = "cm")
```

# Figure 5:
## Panel C: ChiP overlay
```{r}
shockDf = 
  bind_rows(fread(paste0("~/Flank_seq/data/shock_pho4.txt"), header = T) %>% mutate(protein = "pho4"),
            fread(paste0("~/Flank_seq/data/shock_cbf1.txt"), header = T) %>% mutate(protein = "cbf1")) %>%
  tbl_df()

# join NN results with titration data
nnCor = 
  left_join(std_kd %>% select(flank,stdE,protein), shockDf %>% select(flank,ddG,cc,t,protein))

# lm to extract linear equation
slopeDf = foreach(i = c("pho4","cbf1"), .combine = "rbind")%do%{
  iSlice = nnCor %>% filter(protein == i)
  mod = lm(stdE ~ ddG, iSlice)
  tidy(mod) %>% 
    mutate(term = ifelse(term == "(Intercept)","int","slope")) %>%
    select(term,estimate) %>%
    spread(term,estimate) %>%
    mutate(protein = i)
}

nnfitCor = 
  left_join(shockDf,slopeDf) %>%
  mutate(ddG = (ddG*slope)) %>% #only scale, no offset
  select(-int,-slope) %>%
  group_by(cc) %>% 
  mutate(meanval = mean(ddG)) %>%
  ungroup()

overlayDf = 
  left_join(nnfitCor %>% select(flank,ddG,t,cc,protein),chip.gen.full %>% select(flank,pho4,cbf1) %>% gather(protein,occ,pho4,cbf1)) %>% 
  mutate(protein = ifelse(protein == "pho4","Pho4","Cbf1"),
         protein = factor(protein, levels = c("Pho4","Cbf1")))

# circle plot
p1 = ggplot(overlayDf) +
  geom_point(aes(t,cc,color=ddG), size = 0.01) +
  geom_point(aes(t,cc,size=occ,color=ddG)) +
  coord_polar() +
  scale_size_continuous("ChIP enrichment", range = c(0.01, 2.5)) +
  scale_color_gradient2("ddG, kcal/mol",low = "#0057E7", high = "#D62D20", mid = "white") +
  theme(aspect.ratio = 1, 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none") +
  facet_wrap(~ protein, nrow = 1)

ggsave("~/Flanks/Images/f5pC.png", plot = p1, height = 12, width = 12, units = "cm")

# circle plot
pLeg = ggplot(overlayDf) +
  geom_point(aes(t,cc,color=ddG), size = 0.01) +
  geom_point(aes(t,cc,size=occ,color=ddG)) +
  coord_polar() +
  scale_size_continuous("ChIP\nenrichment", range = c(0.01, 2.5)) +
  scale_color_gradient2("ddG\nkcal/mol",low = "#0057E7", high = "#D62D20", mid = "white") +
  theme(aspect.ratio = 1, 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank()) +
  facet_wrap(~ protein, nrow = 1)

ggsave("~/Flanks/Images/f5pC_leg.png", plot = pLeg, height = 12, width = 12, units = "cm")
```
