---
title: "Final figures"
author: "Dan"
date: "8/15/2017"
output: html_document
---

```{r setup, include=FALSE}
source("~/Flank_seq/Figures/setup.R")
```

#Figure 1:
## Panel A: DNA design
## Panel B: Device operation
## Panel C: Sequencing quantitation
## Panel B: Landscape
```{r}
foreach(prot = c("pho4","cbf1"))%dopar%{
  bestFlank =
    countsNN_scaled %>%
    filter(protein == prot) %>%
    select(flank, ddG = scaled_ddG) %>% 
    arrange(ddG) %>%
    slice(1) %>%
    .$flank

  shockDf =
    countsNN_scaled %>%
    filter(protein == prot) %>%
    select(flank, ddG = scaled_ddG) %>% 
    mutate(cc = hdset2(flank,bestFlank)) %>%
    group_by(cc) %>%
    arrange(flank) %>%
    mutate(t = seq(0,360-(360/n()),length.out = n()) * pi / 180) %>%
    ungroup()

  write.table(shockDf, paste0("~/Flank_seq/data/shock_",prot,".txt"), col.names = T, row.names = F, quote = F, sep = "\t")
}

###
prot = "pho4"
###

shockDf = 
  fread(paste0("~/Flank_seq/data/shock_",prot,".txt"), header = T) %>% 
  mutate(t = t/2) %>% #create hemisphere
  tbl_df() %>%
  bind_rows(data.frame(flank = NA, ddG = NA, cc = 11, t = 2*pi))

labelDf = 
  shockDf %>%
  group_by(cc) %>%
  summarise(count = n()) %>%
  ungroup() %>%
  mutate(rolling = cumsum(count), 
         digits = nchar(as.character(count)), 
         filling = paste0(count, "   "))

# circle plot
p1 = ggplot(shockDf) +
  geom_point(aes(t,cc,color=ddG)) +
  coord_polar(start = -pi/2) +
  ylim(c(0,10)) +
  scale_color_gradient2("ddG, kcal/mol",low = "#0057E7", high = "#D62D20", mid = "white") +
  theme(aspect.ratio = 1, 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none",
        legend.text = element_text(angle = 45)) +
  geom_text(data = labelDf %>% filter(cc != 0, cc != 11), aes(pi,cc,label =cc), inherit.aes = F, vjust = 1.5) +
  geom_text(data = labelDf %>% filter(cc != 11), aes(0,cc,label =filling), inherit.aes = F, angle = 45, hjust= 0.9, vjust = 1, color = "gray")

ggsave("~/Flanks/Images/f1pD.png", plot = p1, height = 12, width = 12, units = "cm", dpi = 600)
```

# Figure 2:
## Panel A: Monte Carlo sampling of Spec-seq PWM and ddG

```{r}
## Parameters ###
iVal = unique(c(seq(100,1000,100),seq(1E3,1E4,1E3),seq(1E4,1E5,1E4),seq(1E5,1E6,1E5),seq(1E6,1E7,1E6)))
jVal = 1:10

testCount = lapply(list(iVal,jVal),function (x) x %>% length()) %>%
  unlist() %>%
  prod()
###

# read-in data
readDat2 =
  fread("~/Flank_seq/data/spec_R3_2.txt", header = F) %>% #R3.2 (only from Expt 1)
  tbl_df() %>%
  rename(Seq = V1, True_Bound = V2, True_Unbound = V3, True_Ratio = V4) %>%
  mutate(True_Bound_P = True_Bound / sum(True_Bound),
         True_Unbound_P = True_Unbound / sum(True_Unbound),
         spec_ddG = True_Bound_P / True_Unbound_P, 
         target = Seq) %>%
    separate(target, paste0("Var",1:10),1:9) %>%
  select(Seq,spec_ddG,Var5,Var7:Var10)

predPWM = augment(lm(spec_ddG ~ Var5+Var7+Var8+Var9+Var10, data = readDat2)) %>% 
  tbl_df()

# perform sampling
dsPWM = foreach(i = iVal, .combine = "rbind")%dopar%{ # sampling depth split equally between bound and unbound
  libDepth = i
  dnaSpecies = 1024
  foreach(j = jVal, .combine = "rbind")%do%{ # iterations

    trueRatioDf =
      data.frame(specIndex = readDat2$Seq, trueRatio = readDat2$spec_ddG) %>%
      tbl_df()

    obsRatioDf =
      inner_join(tbl_df(data.frame(sampleBound = sample(trueRatioDf$specIndex, libDepth/2, prob = trueRatioDf$trueRatio, replace = T))) %>%
        count(sampleBound) %>%
        rename(specIndex = sampleBound, bound = n),
        tbl_df(data.frame(sampleUnbound = sample(trueRatioDf$specIndex, libDepth/2, prob = rep(1,dnaSpecies), replace = T))) %>%
        count(sampleUnbound) %>%
        rename(specIndex = sampleUnbound, unbound = n)) %>%
      mutate(obsRatio = bound/unbound)

    ratioDf =
      inner_join(trueRatioDf, obsRatioDf %>% select(-bound, -unbound)) %>%
      mutate(target = specIndex) %>%
      separate(target, paste0("Var",1:10),1:9) %>%
      select(-Var1,-Var2,-Var3,-Var4,-Var6) %>% # determine relative difference
      mutate(relDiff = diff(range(obsRatio))/diff(range(trueRatio)))

    model = tryCatch({lm(obsRatio ~ Var5+Var7+Var8+Var9+Var10, data = ratioDf)},
                     error = function(err) {NA})
    
    modelPred = tryCatch({predict(model, newdata = predPWM %>% select(Var5,Var7:Var10))},
                         error = function(err) {NA})

    rSeq = tryCatch({
      ifelse(cor.test(ratioDf$trueRatio, ratioDf$obsRatio,method = "pearson")$p.value < (0.05/testCount),
             cor.test(ratioDf$trueRatio, ratioDf$obsRatio,method = "pearson")$estimate^2,
             NA)}, 
      error = function(err) {NA})
    
    rPwm = tryCatch({
      ifelse(cor.test(modelPred, predPWM$.fitted,method = "pearson")$p.value < (0.05/testCount) & length(modelPred) == 1024,
             cor.test(modelPred, predPWM$.fitted,method = "pearson")$estimate^2,
             NA)}, 
      error = function(err) {NA})

    data.frame(readTotal = libDepth, 
               meanRead_perSeq = (libDepth/2)/dnaSpecies, 
               iter = j, 
               r2seq = ifelse(is.na(rSeq),0,rSeq), 
               r2pwm = ifelse(is.na(rPwm),0,rPwm), 
               libSize = dnaSpecies, 
               seqObsPct = nrow(obsRatioDf)/dnaSpecies,
               spread = mean(ratioDf$relDiff))
  }
} %>% gather(key,value,c(r2seq,r2pwm)) %>% 
  mutate(acc= ifelse(key == "r2seq", value * seqObsPct, value))

###

#write.table(dsPWM, "~/Flank_seq/data/dspwmSimulation.txt", col.names = T, row.names = F, quote = F)

dsPWM = fread("~/Flank_seq/data/dspwmSimulation.txt", header = T)

colPal = brewer.pal(9,"Set1")
p1 = ggplot(dsPWM %>% filter(readTotal <= 1E6)) +
  presentation +
  ylim(c(0,1)) +
  geom_point(aes(readTotal,acc,group = interaction(readTotal,key),color=key), 
             stat = "summary", 
             fun.y = "median",
             size = 0.25) +
  geom_line(aes(readTotal,acc,color = key), 
            stat="summary",
            fun.y = "median", 
            size = 0.25) +
  scale_x_log10(breaks = 10^(2:6), labels = expression(bold("10"^"2"),bold("10"^"3"),bold("10"^"4"),bold("10"^"5"),bold("10"^"6"))) +
  scale_color_manual("", 
                     values = colPal[1:2],
                     labels = expression(bold("mononucleotide\nadditive model"),bold(paste("individual ",Delta,Delta,"G"))),
                     guide = guide_legend(label.vjust = 0.5,
                                          label.hjust = 0)) +
  theme(aspect.ratio = 1/2,
        legend.key.size = unit(0.3, "cm")) +
  ylab(("Median\nnormalized accuracy")) +
  xlab("Sequencing depth, read counts")

ggsave("~/Flanks/Images/f2pA.svg", plot = p1, height = 3, width = 8.7, units = "cm")
```

## Panel B: Simulate assay output as function of defined parameters.

```{r, eval=FALSE}
### Parameters ###
# mVal = c(0.25,0.5,1:5)
# nVal = 10^(2:6)
# iVal = 10^(3:8)
# jVal = 1:10
# 
# testCount = lapply(list(mVal,nVal,iVal,jVal),function (x) x %>% length()) %>%
#   unlist() %>%
#   prod()
# 
# typeCor = "pearson"
# ###
# 
# returnDf =
#   foreach(m = mVal, .combine = "rbind")%do%{ # ddG = 0.25-5.0 corresponds to 1.5- to 4590- fold change
#     foreach(n = nVal, .combine = "rbind")%do%{ # library size
#       gibbsDiff = m
#       dnaSpecies = n
#       dnaRatios = sapply(rnorm(dnaSpecies, mean = 0, sd = gibbsDiff/6), function(x) exp(-x/.593)) #RT at 298K
#       foreach(i = iVal, .combine = "rbind")%do%{ # sampling depth
#         libDepth = i
#         print(paste(m,n,i))
#         foreach(j = jVal, .combine = "rbind")%dopar%{ # 10 iterations
#           trueRatioDf =
#             tbl_df(data.frame(specIndex = 1:dnaSpecies, trueRatio = dnaRatios)) %>%
#             mutate(prob = trueRatio/sum(trueRatio))
#           obsRatioDf =
#             inner_join(tbl_df(data.frame(sampleBound = sample(trueRatioDf$specIndex, libDepth/2, prob = trueRatioDf$prob, replace = T))) %>%
#               count(sampleBound) %>%
#               rename(specIndex = sampleBound, bound = n),
#               tbl_df(data.frame(sampleUnbound = sample(trueRatioDf$specIndex, libDepth/2, prob = rep(1/dnaSpecies,dnaSpecies), replace = T))) %>%
#               count(sampleUnbound) %>%
#               rename(specIndex = sampleUnbound, unbound = n)) %>%
#             mutate(obsRatio = bound/unbound)
#           ratioDf =
#             inner_join(trueRatioDf %>% select(-prob), obsRatioDf %>% select(-bound, -unbound))
#           rSqd =
#             tryCatch({ifelse(cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$p.value < (0.05/testCount), cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$estimate, NA)}, error = function(err) {NA})
# 
#           data.frame(readTotal = libDepth, meanRead_perSeq = (libDepth/2)/dnaSpecies, iter = j, r2 = ifelse(is.na(rSqd),0,rSqd), ddG = gibbsDiff, libSize = dnaSpecies, seqObsPct = nrow(obsRatioDf)/dnaSpecies) %>%
#             mutate(r2 = ifelse(typeCor == "pearson", r2^2, r2),
#                    r2 = r2 * seqObsPct)
#         }
#       }
#     }
#   }
# 
# write.table(returnDf, "~/Flank_seq/data/modelSimulation.txt", col.names = T, row.names = F, quote = F)
###

returnDf2 =
  fread("~/Flank_seq/data/modelSimulation.txt", header = T) %>%
  mutate(r2 = r2/seqObsPct) %>% 
  gather(key,value,r2,seqObsPct) %>% 
  tbl_df() %>%
  mutate(key = ifelse(key == "r2", 1, 2),
         key = factor(key, labels = c(TeX("$\\textbf{Pearson's}$ $\\textbf{\\textit{r}}$$\\mathbf{^2}$"), TeX("$\\textbf{fract. obs.}$"))),
         readTotal = log10(readTotal),
         libSize = log10(libSize),
         libSize = factor(libSize, labels = c(
           expression(bold(paste("lib. size = 10"^"2"))),
           expression(bold(paste("lib. size = 10"^"3"))),
           expression(bold(paste("lib. size = 10"^"4"))),
           expression(bold(paste("lib. size= 10"^"5"))),
           expression(bold(paste("lib. size = 10"^"6")))
         )))

colPal = colorRampPalette(brewer.pal(9,"YlOrRd"))(10)
p1 = ggplot(returnDf2 ) +
  presentation +
  geom_line(aes(ddG, value, color = as.factor(readTotal), group = as.factor(readTotal)),stat = "summary",fun.y = "median", na.rm = T, size = 0.25) +
  geom_point(aes(ddG, value, color = as.factor(readTotal), group = as.factor(readTotal)),stat = "summary",fun.y = "median", na.rm = T, size = 0.25) +
  theme(aspect.ratio = 1, 
        axis.text.x = element_text(vjust = 0.5),
        legend.key.size = unit(0.3, "cm")) +
  facet_grid(libSize ~ key, scales = "free", labeller = "label_parsed") +
  ylim(c(0,1)) +
  xlab(expression(bold(paste(Delta,Delta,"G, kcal/mol")))) +
  ylab("Median") +
  scale_color_manual("Sequencing depth", 
                     values = colPal[(length(colPal)-6):length(colPal)], 
                     labels = expression(bold("10"^"3"),bold("10"^"4"),bold("10"^"5"),bold("10"^"6"),bold("10"^"7"),bold("10"^"8")))

ggsave("~/Flanks/Images/f2pB.svg", plot = p1, height = 10.5, width = 8.7, units = "cm")
```

## Panel C: Flank library depth simultion
```{r}
# FYI spec-seq = 5.4kt (3.2 kcal/mol) affinity range
# readDat =
#   fread("~/Flank_seq/data/Stormo_All_Expts.txt", header = T) %>%
#   tbl_df() %>%
#   rename(Seq = Sequence, True_Bound = Bound, True_Unbound = Unbound, True_Ratio = Bound.Unbound) %>%
#   mutate(index = c(rep(1,256), rep(2,256), rep(3,1024), rep(4,1024))) %>% #R2,R4,R3.1,R3.2 (only from Expt 1)
#   group_by(index) %>%
#   mutate(True_Bound_P = True_Bound / sum(True_Bound),
#          True_Unbound_P = True_Unbound / sum(True_Unbound),
#          spec_ddG = True_Bound_P / True_Unbound_P) %>%
#   ungroup() %>%
#   mutate(Seq = 1:nrow(.)) %>%
#   select(Seq,spec_ddG) %>%
#   right_join(.,data.frame(Seq = 1:(4^10)) %>%
#   mutate(ref_ddG = sapply(rnorm(4^10, mean = 0, sd = 1/6), function(x) exp(-x/.593))))
# 
# ### Parameters ###
# iVal = unique(c(seq(1E1,1E2,1E1),seq(1E2,1E3,1E2),seq(1E3,1E4,1E3),seq(1E4,1E5,1E4),seq(1E5,1E6,1E5),seq(1E6,1E7,1E6),seq(1E7,1E8,1E7),seq(1E8,1E9,1E8)))
# jVal = 1:10
# kVal = c("spec_ddG","ref_ddG")
# 
# testCount = lapply(list(iVal,jVal,kVal),function (x) x %>% length()) %>%
#   unlist() %>%
#   prod()
# ###
# 
# dsDat = foreach(i =iVal, .combine = "rbind")%do%{ # sampling depth split equally between bound and unbound
#         print(i)
#         foreach(k = kVal, .combine = "rbind")%do%{ # spec-seq or theoretical normal
#           foreach(j = jVal, .combine = "rbind")%dopar%{ # iterations
# 
#             libDepth = i
# 
#             dnaSpecies = length(readDat[,k] %>% unlist(use.names=F) %>% na.omit())
# 
#             trueRatioDf =
#               tbl_df(data.frame(specIndex = 1:dnaSpecies, trueRatio = readDat[,k] %>% unlist(use.names=F) %>% na.omit()))
# 
#             obsRatioDf =
#               inner_join(tbl_df(data.frame(sampleBound = sample(trueRatioDf$specIndex, libDepth/2, prob = trueRatioDf$trueRatio, replace = T))) %>%
#                 count(sampleBound) %>%
#                 rename(specIndex = sampleBound, bound = n),
#                 tbl_df(data.frame(sampleUnbound = sample(trueRatioDf$specIndex, libDepth/2, prob = rep(1,dnaSpecies), replace = T))) %>%
#                 count(sampleUnbound) %>%
#                 rename(specIndex = sampleUnbound, unbound = n)) %>%
#               mutate(obsRatio = bound/unbound)
# 
#             ratioDf =
#               inner_join(trueRatioDf, obsRatioDf %>% select(-bound, -unbound))
# 
#             rSqd = tryCatch({ifelse(cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$p.value < (0.05/testCount),cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = typeCor)$estimate,NA)}, error = function(err) {NA})
# 
#             error = tryCatch({ifelse(cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = "pearson" )$p.value < (0.05/testCount),cor.test(ratioDf$trueRatio, ratioDf$obsRatio, method = "pearson")$estimate^2,NA)}, error = function(err) {NA})
# 
#             data.frame(readTotal = libDepth, meanRead_perSeq = (libDepth/2)/dnaSpecies, iter = j, r2 = ifelse(is.na(rSqd),0,rSqd), err = ifelse(is.na(error),1,1-error),libSize = dnaSpecies, seqObsPct = nrow(obsRatioDf)/dnaSpecies) %>%
#               mutate(type = k, 
#                      r2 = ifelse(typeCor == "pearson", r2^2, r2),
#                      r2 = r2 * seqObsPct)
#           }
#         }
# } %>% tbl_df()
# 
# write.table(dsDat, "~/Flank_seq/data/downsampleSimulation.txt", col.names = T, row.names = F, quote = F)
###

dsDat = fread("~/Flank_seq/data/downsampleSimulation.txt", header = T) %>% 
  filter(type == "ref_ddG") %>% 
  mutate(r2 = r2/seqObsPct) %>% 
  gather(key,value,seqObsPct,r2) %>% 
  mutate(key = factor(key, levels = c("seqObsPct","r2")))

#plot
colPal = brewer.pal(9,"Set1")
p1 = ggplot(dsDat) +
  presentation +
  geom_rect(aes(xmin=5, xmax=50, ymin=-0.05, ymax=1.05), fill="cyan", alpha=0.1) +
  geom_line(aes(meanRead_perSeq, value, color = key), stat = "summary", fun.y = "median", size = 0.25) +
  geom_point(aes(meanRead_perSeq, value, color = key), stat = "summary", fun.y = "median", size = 0.25) +
  theme(aspect.ratio = 1/2,
        legend.text.align = 0,
        legend.key.size = unit(0.3, "cm")) +
  xlab("Mean reads per sequence") +
  ylab("Median") +
  scale_x_log10(limits = c(0.1,1000),
                breaks = c(0.1,1,10,100,1000),
                labels = expression(bold("10"^"-1"),bold("10"^"0"),bold("10"^"1"),bold("10"^"2"),bold("10"^"3"))) +
  scale_color_manual(labels = c(expression("fract. obs.",
                                           bold(paste("Pearson's ",italic(r)^"2")))),
                     "", 
                     values = colPal[1:2])

ggsave("~/Flanks/Images/f2pC.svg", plot = p1, height = 3, width = 8.7, units = "cm")
ggsave("~/Flanks/Images/f2pC.png", plot = p1, height = 3, width = 8.7, units = "cm")
ggsave("~/Flanks/Images/f2pC.eps", plot = p1, height = 3, width = 8.7, units = "cm")

dsDat %>% 
  tbl_df() %>% 
  filter(key == "r2", readTotal <= 10000000) %>% 
  na.omit() %>% 
  arrange(desc(value))
```

# Figure 3:
## Panel A: Composite additive model of Pho4 and Cbf1
```{r}
# additive model plots
outAdd = addMotif(countsNN_scaled %>% 
  select(flank, ddG = scaled_ddG, protein) %>% 
  mutate(target = flank, rep = 1) %>% 
  separate(target, paste0("X", sprintf("%02d", 1:10)),1:9), "ddG") %>% 
  mutate(monuc.value = factor(monuc.value, levels = c("T","G","C","A"))) %>% 
  mutate(protein = ifelse(protein == "pho4","Pho4","Cbf1"),
         protein = factor(protein, levels = c("Pho4","Cbf1"))) %>%
  left_join(., data.frame(monuc.value = c("A","C","G","T"), colPal = c("#55C341","#3036DF","#F0E328","#EF1717"))) %>% 
  mutate(colPal = as.character(colPal))

# composite textmap
p1 = ggplot(outAdd,aes(monuc.pos, monuc.meanval)) +
  presentation +
  geom_text(aes(label = monuc.value, color = monuc.value), size = 2.5) +
  theme(aspect.ratio = 1, 
        legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45,
                                   hjust = 1)) +
  ylab(expression(bold(paste("Mean ", Delta, Delta,"G, kcal/mol")))) +
  xlab("Flank position") +
  scale_y_reverse() +
  facet_grid(.~protein, scales = "free") +
  scale_color_manual(values = c("#55C341","#3036DF","#F0E328","#EF1717"))

ggsave("~/Flanks/Images/f3pA.svg", plot = p1, height = 4.5, width = 8.7, units = "cm")
ggsave("~/Flanks/Images/f3pA.png", plot = p1, height = 4.5, width = 8.7, units = "cm")
```

## Panel B: PWM from literature of Pho4 and Cbf1
size = 3 x 8.7 cm
```{r}
pwmList = list("Pho4" = fread("~/Flank_seq/data/pwm_pho4.txt", header = F) %>% as.matrix(),
               "Cbf1" = fread("~/Flank_seq/data/pwm_cbf1.txt", header = F) %>% as.matrix())
row.names(pwmList[["Pho4"]]) = c("A","C","G","T")
row.names(pwmList[["Cbf1"]]) = c("A","C","G","T")

cs1 = make_col_scheme(chars=c('A', 'T', 'C', 'G'), 
                      groups=c('gr1', 'gr2', 'gr3', 'gr4'), 
                      cols=c("#55C341","#EF1717","#3036DF","#F0E328"))

# Generate sequence logo
p1 = ggplot() + 
  geom_logo(pwmList[["Pho4"]],col_scheme=cs1) +
  presentation +
  theme(legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        aspect.ratio = 1/2,
        axis.text.x = element_text(angle = 45,
                                   hjust = 1)) +
  scale_x_continuous(breaks=1:8, labels=c("-1",rep("",6),"1"))

p2 = ggplot() +
  geom_logo(pwmList[["Cbf1"]],col_scheme=cs1) +
  presentation +
  theme(legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        aspect.ratio = 1/2,
        axis.text.x = element_text(angle = 45,
                                   hjust = 1)) +
  scale_x_continuous(breaks=1:8, labels=c("-2","-1",rep("",6)))

pp = cowplot::plot_grid(p1,p2,nrow=1)

ggsave("~/Flanks/Images/f3pB.svg", plot = pp, height = 3, width = 8.7, units = "cm")
```

# Figure 4:
## Panel A: B2 bomber of Pho4 + Cbf1

## Panel B: Pho4 + Cbf1 LASSO regression
```{r}
# Specify protein
protVal = "cbf1"

# Define number of interacting positions
intClass = 2

# Determine the number of terms at each model layer
modelTerms = 
  data.frame(intClass = 1:9) %>% 
  tbl_df() %>%
  mutate(intSet = 4^intClass, intPos = choose(10,intClass), intTerm = intSet * intPos)

# Determine positions of interaction
listDfPos = foreach(i = 1:intClass)%do%{
  expand.grid(rep(list(1:10),i)) %>%
  filter_(.dots = if(i > 1) {paste0("Var",2:i,">Var",1:(i-1))}) # remove redundancy
  }
  
###

# # Subset and encode dataframe for LASSO input 
# doMC::registerDoMC(cores = 40)
# 
# summDat = countsNN_scaled %>% filter(protein == protVal) %>% select(flank, protein, ddG = scaled_ddG)
# testDat = encodeTerms(summDat, 10000, 2, listDfPos)
# 
# # Classify predictor and response elements
# options(contrasts=c(ordered='contr.Dummy', unordered='contr.Dummy')) # no intercept
# x <- model.matrix( ~ .-1, testDat %>% select(contains("Var")))
# y <- testDat %>% .$ddG
# 
# # Perform LASSO regression
# fit.lasso=glmnet(x,y,alpha=1, intercept=FALSE, nlambda = 150) # no intercept, alpha = 1 for lasso, = 0 for ridge
# fitLasso = tidy(fit.lasso)
# write.table(fitLasso, paste0("~/Flank_seq/data/lassoFit_",protVal,".txt"), row.names = F, col.names = T, quote = F, sep = "\t")
# 
# # 10-fold cross validation
# cv.lasso=cv.glmnet(x,y, intercept=FALSE, parallel = T)
# cvLasso = tidy(cv.lasso)
# write.table(cvLasso, paste0("~/Flank_seq/data/lassoCV_",protVal,".txt"), row.names = F, col.names = T, quote = F, sep = "\t")
# 
# # # Extract coefficients
# dfVar =
#   coef(cv.lasso) %>%
#   as.data.frame.matrix() %>%
#   tbl_df() %>%
#   mutate(vars = coef(cv.lasso) %>% rownames()) %>%
#   rename(coef = `1`) %>%
#   filter(vars != "(Intercept)") %>%
#   separate(vars, c("pos","term"),"[.]") %>%
#   group_by(coef,pos,term) %>%
#   mutate(row = substring(pos,4,nchar(pos)) %>% as.integer(),
#          flank = convertFlank(term,row,listDfPos)) %>%
#   ungroup() %>%
#   select(flank, coef) %>%
#   group_by(flank) %>%
#   mutate(space = (gregexpr("[A,C,G,T]", flank) %>% unlist() %>% diff() - 1) %>% sum(),
#          elements = gregexpr("[A,C,G,T]", flank) %>% unlist() %>% length()) %>%
#   ungroup()
# write.table(dfVar, paste0("~/Flank_seq/data/dfVar_",protVal,".txt"), row.names = F, col.names = T, quote = F, sep = "\t")

###

# LASSO plot
fitLasso = bind_rows(fread(paste0("~/Flank_seq/data/lassoFit_pho4.txt"), header = T) %>% mutate(protein = "pho4"),
                     fread(paste0("~/Flank_seq/data/lassoFit_cbf1.txt"), header = T) %>% mutate(protein = "cbf1")) %>% 
  separate(term,c("variable","base"),sep = "[.]") %>%
  mutate(dfrow = substring(variable, 4, nchar(variable)) %>% as.integer()) %>%
  group_by(estimate,lambda,dev.ratio,step) %>%
  mutate(flank = convertFlank(base,dfrow,listDfPos)) %>%
  ungroup()

lassoPlot = foreach(j = c("pho4","cbf1"), .combine = "rbind")%do%{
  jSlice = fitLasso %>% filter(protein == j)
  foreach(i = unique(jSlice %>% .$flank), .combine = "rbind")%do%{
    allLambda = jSlice %>% select(lambda) %>% distinct()
    iSlice = jSlice %>% select(lambda,estimate,flank,variable,base) %>% filter(flank == i)
    flankName = iSlice %>% .$flank %>% unique()
    varName = iSlice %>% .$variable %>% unique()
    baseName  = iSlice %>% .$base %>% unique()
    left_join(allLambda,iSlice %>% select(lambda,estimate), by = "lambda") %>%
      mutate(estimate = ifelse(is.na(estimate),0,estimate), 
             flank = flankName,
             flank = paste0(substr(flank,1,5),"_",substr(flank,6,10)),
             variable = varName, 
             base = baseName,
             protein = j,
             protein = ifelse(protein == "pho4","Pho4","Cbf1"),
             protein = factor(protein, levels = c("Pho4","Cbf1")))
  }
} %>% mutate(type = nchar(base),
             feature = ifelse(type == 1,"mononucleotide","dinucleotide"),
             feature = factor(feature, levels = c("mononucleotide","dinucleotide")))

labelDf2 = 
  lassoPlot %>%
  filter(estimate != 0) %>% 
  group_by(protein, feature, flank) %>%
  mutate(maxLamdha = max(lambda)) %>% 
  filter(maxLamdha == lambda) %>% 
  mutate(signGroup = sign(estimate)) %>% 
  group_by(protein, feature, signGroup) %>% 
  mutate(rank = dense_rank(lambda)) %>%
  filter(rank >= max(rank)-3) %>% 
  mutate(rank = dense_rank(lambda),
         id = row_number(),
         colVal = signGroup) %>%
  filter(id <= 3) %>% 
  ungroup() %>% 
  select(protein,flank,id,colVal,feature) %>% 
  left_join(.,data.frame(id = c(1:3,3:1), yPos = c(0.6,0.4,0.2,-0.7,-0.9,-1.1), colVal = c(1,1,1,-1,-1,-1)))

lassoPlot2 = 
  left_join(lassoPlot, labelDf2) %>% 
  mutate(colVal = ifelse(is.na(colVal),0,colVal))

p1 = ggplot(lassoPlot2) +
  presentation +
  geom_line(aes(lambda,estimate,group=interaction(variable,base), color = colVal), size = 0.25) +
  theme(legend.position = "none") +
  ylab("Variable coefficient") +
  xlab(expression(bold(paste("Penalty coefficient, ", lambda)))) +
  scale_x_reverse() +
  geom_label(data = labelDf2,  
             aes(x = 0.22, y = yPos, label = flank, color = colVal), 
             inherit.aes = F, 
             size = 1.75, 
             label.padding = unit(0.1, "lines"),
             label.r = unit(0.05, "lines")) +
  scale_color_gradient2(high = "red", mid = "gray", low = "blue") +
  facet_wrap(~ protein+feature, nrow = 1) +
  ylim(c(-1.2,0.7))

ggsave("~/Flanks/Images/f4pB.svg", plot = p1, height = 5, width = 17, units = "cm")
```

# Figure 5:
## Panel A: Analysis flow chart
## Panel B: Interpretation scatter
## Panel C: Pareto plot
## Panel D: ChiP overlay
```{r}
shockDf = 
  bind_rows(fread(paste0("~/Flank_seq/data/shock_pho4.txt"), header = T) %>% mutate(protein = "pho4"),
            fread(paste0("~/Flank_seq/data/shock_cbf1.txt"), header = T) %>% mutate(protein = "cbf1")) %>%
  tbl_df()

# join NN results with titration data
nnCor = 
  left_join(std_kd %>% select(flank,stdE,protein), shockDf %>% select(flank,ddG,cc,t,protein))

# lm to extract linear equation
slopeDf = foreach(i = c("pho4","cbf1"), .combine = "rbind")%do%{
  iSlice = nnCor %>% filter(protein == i)
  mod = lm(stdE ~ ddG, iSlice)
  tidy(mod) %>% 
    mutate(term = ifelse(term == "(Intercept)","int","slope")) %>%
    select(term,estimate) %>%
    spread(term,estimate) %>%
    mutate(protein = i)
}

nnfitCor = 
  left_join(shockDf,slopeDf) %>%
  mutate(ddG = (ddG*slope)) %>% #only scale, no offset
  select(-int,-slope) %>%
  group_by(cc) %>% 
  mutate(meanval = mean(ddG)) %>%
  ungroup()

overlayDf = 
  left_join(nnfitCor %>% select(flank,ddG,t,cc,protein),chip.gen.full %>% select(flank,pho4,cbf1) %>% gather(protein,occ,pho4,cbf1)) %>% 
  mutate(protein = ifelse(protein == "pho4","Pho4","Cbf1"),
         protein = factor(protein, levels = c("Pho4","Cbf1"))) %>% 
  group_by(protein) %>% 
  mutate(relOcc = occ/max(occ, na.rm = T)) %>% 
  ungroup()

# circle plot (no ChIP)
p1 = ggplot(overlayDf) +
  geom_point(aes(t,cc,color=ddG), size = 0.01) +
  coord_polar() +
  scale_size_continuous("ChIP enrichment", range = c(0.01, 2.5)) +
  scale_color_gradient2("ddG, kcal/mol",low = "#0057E7", high = "#D62D20", mid = "white") +
  theme(aspect.ratio = 1, 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none") +
  facet_wrap(~ protein, nrow = 1)

ggsave("~/Flanks/Images/f5pC1.png", plot = p1, height = 12, width = 12, units = "cm", dpi = 600)

# circle plot (only ChIP)
p1 = ggplot(overlayDf) +
  theme_bw() +
  geom_point(aes(t,cc,size=relOcc,color=ddG)) +
  coord_polar() +
  scale_size_continuous("ChIP enrichment", range = c(0.01, 2.5)) +
  scale_color_gradient2("ddG, kcal/mol",low = "#0057E7", high = "#D62D20", mid = "white") +
  theme(aspect.ratio = 1, 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none",
        panel.border = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(size = 0.1, color = "gray40")) +
  scale_y_continuous(breaks = 0:10) +
  facet_wrap(~ protein, nrow = 1)

ggsave("~/Flanks/Images/f5pC2.png", plot = p1, height = 12, width = 12, units = "cm", dpi = 600)

# legend print
pLeg = ggplot(overlayDf) +
  geom_point(aes(t,cc,color=ddG), size = 0.01) +
  geom_point(aes(t,cc,size=relOcc,color=ddG)) +
  coord_polar() +
  scale_size_continuous(expression(bold("Relative ChIP\nenrichment")), range = c(0.01, 2.5)) +
  scale_color_gradient2(expression(bold(paste(Delta,Delta,"G, kcal/mol"))),low = "#0057E7", high = "#D62D20", mid = "white") +
  theme(aspect.ratio = 1,
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank()) +
  facet_wrap(~ protein, nrow = 1)

ggsave("~/Flanks/Images/f5pCleg.png", plot = pLeg, height = 12, width = 12, units = "cm", dpi = 600)
```




########
########
########

# Tyler figures


# Comparisons between mono- and dinucleotide models and NN predictions

Setup for the figures: make models with NN-predicted data and save to model_outputs.txt. This step is usually run on Zion.

```{r eval=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
library(tidyr)
library(broom)

# Function to create models and extract predicted values
create_models <- function(data) {
  
  # Get the protein name
  dataset_name <- data$protein[1]
  
  # Select remove proetin column
  data <- select(data, -protein)
  
  # Make the mononucleotide model
  model <- lm(scaled_ddG ~ ., data = data)
  mono_nuc <- augment(model, data)
  
  mono_nuc$model <- "mono-nucleotide"
  
  # Make nearest neighbor dinucleotide model
  model <- lm(scaled_ddG ~ X01 +X02 + X03 + X04 + X05 +
                X06 + X07 + X09 + X10 +
                X01:X02 + X02:X03 + X03:X04 + X04:X05 +
                X06:X07 + X07:X08 + X08:X09 + X09:X10, data = data)
  nearest_n <- augment(model, data)
  
  nearest_n$model <- "nearest neighbor"
  
  # Make full dinucleotide model
  model <- lm(scaled_ddG ~ .*., data = data)
  di_nuc <- augment(model, data)
  
  di_nuc$model <- "di-nucleotide"
  
  # Join and output all the data
  model_preds <- bind_rows(mono_nuc, nearest_n, di_nuc)
  
  model_preds$dataset <- dataset_name
  
  return(model_preds)
}

# Read in and ajust the NN predicted ddG values
data <- fread("~/Flank_seq/data/scaled_nn_preds.txt") %>% 
  select(flank, protein, scaled_ddG) %>% 
  separate(flank, into=paste0("X", sprintf("%02d", 1:10)), sep = 1:9)

# Generate models and get predicted values for all models for all preoteins
model_outputs <- data %>% 
  group_by(protein) %>% 
  do(create_models(.))

# Write model results to text file
write.table(model_outputs, "~/model_outputs.txt", row.names = FALSE, col.names = TRUE, sep = "\t", quote = FALSE)
```

## Read in the preprocessed data
```{r}
# Read in model output data and put into simple model form
model_outputs <- fread("~/Google Drive/Flanks_data/model_outputs.txt")

simple_model_data <- model_outputs %>% 
  ungroup() %>% 
  # unite(flank, X01:X10, sep = "") %>%
  rename(ddG = scaled_ddG) %>% 
  select(flank, ddG, .fitted, model, dataset) %>% 
  mutate(model = ifelse(model == "mono-nucleotide",
                        "mononucleotide", ifelse(model == "nearest neighbor",
                                                 "nearest neighbor", "dinucleotide")),
         model = factor(model, levels = c("mononucleotide", "nearest neighbor", "dinucleotide")),
         dataset = factor(dataset, levels = c("Pho4", "Cbf1")))

```

## Mononucleotide models 
```{r}
library(latex2exp)

ggplot(simple_model_data %>% filter(model == "mononucleotide"), aes(x = .fitted, y = ddG)) +
  geom_hex(bins = 50) +
  scale_fill_gradient2(name = "Count", trans = "log",
                       low = "black", mid = "red", high = "yellow", midpoint = 6,
                       breaks = c(1, 10, 100, 1000, 10000)) +
  facet_grid(model~dataset) +
  coord_fixed() +
  xlab(TeX("\\textbf{Linear model predicted $\\Delta \\Delta G$, kcal/mol}")) +
  ylab(TeX("\\textbf{Neural network predicted $\\Delta \\Delta G$, kcal/mol}")) +
  presentation +
  theme(legend.key.size = unit(0.25, "cm"))

ggsave("~/Flanks/Images/f3pX.svg", width = 10, height = 6,
       units = "cm")
```

## Dinucleotide models
```{r}
ggplot(simple_model_data %>% filter(model != "mononucleotide"), aes(x = .fitted, y = ddG)) +
  geom_hex(bins = 50) +
  scale_fill_gradient2(name = "Count", trans = "log",
                       low = "black", mid = "red", high = "yellow", midpoint = 6,
                       breaks = c(1, 10, 100, 1000, 10000)) +
  facet_grid(model~dataset) +
  coord_fixed() +
  xlab(TeX("\\textbf{Linear model prediction $\\Delta \\Delta G$, kcal/mol}")) +
  ylab(TeX("\\textbf{Neural network prediction $\\Delta \\Delta G$, kcal/mol}")) +
  presentation +
  theme(legend.key.size = unit(0.25, "cm"))

ggsave("~/Flanks/Images/f4pX.svg", width = 10, height = 6,
       units = "cm")
```


```{r}
# Total r^2
simple_model_data %>%
  group_by(model, dataset) %>%
  summarize(r2 = cor(ddG, .fitted)^2)

# Held-out r^2
training_data <- fread("~/Google Drive/Flanks_data/train_seqs.txt", header = FALSE) %>% 
  rename(flank = V1)

simple_model_data %>%
  anti_join(., training_data) %>% 
  group_by(model, dataset) %>%
  summarize(r2 = cor(ddG, .fitted)^2)
```


# B2 Bomber Plot

```{r}
library(data.table)
library(dplyr)
library(ggplot2)
library(tidyr)
library(broom)

# Read in the count data from our experiments
data <- fread("~/Google Drive/Flanks_data/scaled_nn_preds.txt") %>% 
  separate(flank, into = paste0("X", sprintf("%02d", 1:10)), sep = 1:9, remove = FALSE)

# Define the function to get the mean residual multiplied by the scalar
get_effect <- function(col1, col2, data) {
  colnames(data)[which(colnames(data) == col1)] <- "nt1"
  colnames(data)[which(colnames(data) == col2)] <- "nt2"

  output <- data %>%
    group_by(nt1, nt2) %>%
    summarize(ddG = mean(.resid)) %>%
    mutate(pos1 = col1, pos2 = col2)

  return(output)
}

get_mean_effects <- function(data) {
  colnames <- colnames(data)[1:10]
  combos <- expand.grid(pos1 = colnames, pos2 = colnames) %>% 
    mutate(num1 = as.numeric(gsub("X", "", pos1)),
           num2 = as.numeric(gsub("X", "", pos2))) %>% 
    filter(num1 < num2) %>% 
    select(pos1, pos2)
  
  output <- combos %>%
    group_by(pos1, pos2) %>%
    do(get_effect(.$pos1[1], .$pos2[1], data))
}

# Make a function to plot the mean dinuc residual and save to a file
make_plot <- function(data, data_string, kd_data){
  
  # Handle the presentation style
  presentation <-  theme_bw() + 
    theme(plot.title = element_text(hjust = 0.5,
                                    family = "Arial",
                                    face = "bold",
                                    size = 24),
          axis.title = element_text(family = "Arial",
                                    face = "bold",
                                    size = 18),
          axis.text = element_text(family = "Arial",
                                   face = "bold",
                                   size = 14),
          strip.text = element_text(family = "Arial",
                                    face = "bold",
                                    size = 12),
          legend.title = element_text(family = "Arial",
                                      face = "bold",
                                      size = 14),
          legend.text = element_text(family = "Arial",
                                     face = "bold",
                                     size = 12))
  
  # Subset the data to the protein of interest
  new_data <- data %>%
    filter(protein == data_string) %>% 
    select(X01:X10, scaled_ddG)
  
  # Make the linear model
  model <- lm(scaled_ddG ~ ., data = new_data)
  
  # Extract the fitted and resid values
  model_results <- augment(model) %>% 
    select(X01:X10, .resid, .fitted)
  
  # Titration comparison
  plot_data <- get_mean_effects(model_results) %>% 
    ungroup() %>% 
    mutate(pos1 = as.numeric(gsub("X", "", pos1)) - 6,
           pos2 = as.numeric(gsub("X", "", pos2)) - 6) %>% 
    mutate(pos1 = pos1 + (pos1 >= 0),
           pos2 = pos2 + (pos2 >= 0))
  
  plot1 <- ggplot(plot_data) +
    facet_grid(pos2~pos1, space = "free") +
    geom_tile(aes(nt1, nt2, fill=ddG)) +
    presentation +
    theme(aspect.ratio = 1) +
    # scale_fill_gradient2(low = "#0057E7", high = "#D62D20", mid = "white") +
    scale_fill_gradient2(low = "#0057E7", high = "#D62D20", mid = "white", limits = c(-0.51, 0.51)) +
    theme(axis.text.y = element_text(angle = 270, color = "black", hjust = 0.5),
          axis.text.x = element_text(angle = 0, color = "black"),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(size = 2)) +
    xlab("") +
    ylab("")
  
  return(plot1)
  
  ggsave(paste0("~/Flank_seq/Figures/Final_Figures/Figure_4/", data_string, "_dinuc_mean_ddg.svg"),
         height = 10, width = 10, units = "in")
  
}

make_plot(data, "Pho4", std_kd)
test <- make_plot(data, "Cbf1", std_kd)

test
```





